// @hookform/resolvers@3.10.0/joi/src/__tests__/__fixtures__/data.ts
import * as Joi from 'joi';

import { Field, InternalFieldName } from 'react-hook-form';

export const schema = Joi.object({
  username: Joi.string().alphanum().min(3).max(30).required(),
  password: Joi.string()
    .pattern(new RegExp('.*[A-Z].*'), 'One uppercase character')
    .pattern(new RegExp('.*[a-z].*'), 'One lowercase character')
    .pattern(new RegExp('.*\\d.*'), 'One number')
    .pattern(
      new RegExp('.*[`~<>?,./!@#$%^&*()\\-_+=&quot;\&apos;|{}\\[\\];:\\\\].*&apos;),
      &apos;One special character&apos;,
    )
    .min(8)
    .required(),
  repeatPassword: Joi.ref(&apos;password&apos;),
  accessToken: [Joi.string(), Joi.number()],
  birthYear: Joi.number().integer().min(1900).max(2013),
  email: Joi.string().email({
    minDomainSegments: 2,
    tlds: { allow: [&apos;com&apos;, &apos;net&apos;] },
  }),
  tags: Joi.array().items(Joi.string()).required(),
  enabled: Joi.boolean().required(),
  like: Joi.array()
    .items(
      Joi.object({ id: Joi.number(), name: Joi.string().length(4).regex(/a/) }),
    )
    .optional(),
});

interface Data {
  username: string;
  password: string;
  repeatPassword: string;
  accessToken?: number | string;
  birthYear?: number;
  email?: string;
  tags: string[];
  enabled: boolean;
  like: { id: number; name: string }[];
}

export const validData: Data = {
  username: &apos;Doe&apos;,
  password: &apos;Password123_&apos;,
  repeatPassword: &apos;Password123_&apos;,
  birthYear: 2000,
  email: &apos;john@doe.com&apos;,
  tags: [&apos;tag1&apos;, &apos;tag2&apos;],
  enabled: true,
  like: [
    {
      id: 1,
      name: &apos;name&apos;,
    },
  ],
};

export const invalidData = {
  password: &apos;___&apos;,
  email: &apos;&apos;,
  birthYear: &apos;birthYear&apos;,
  like: [{ id: &apos;z&apos;, name: &apos;r&apos; }],
};

export const fields: Record<InternalFieldName, Field['_f']> = {
  username: {
    ref: { name: 'username' },
    name: 'username',
  },
  password: {
    ref: { name: 'password' },
    name: 'password',
  },
  email: {
    ref: { name: 'email' },
    name: 'email',
  },
  birthday: {
    ref: { name: 'birthday' },
    name: 'birthday',
  },
};
