// tailwind-merge@2.6.0/src/lib/default-config.ts
import { fromTheme } from './from-theme'
import { Config, DefaultClassGroupIds, DefaultThemeGroupIds } from './types'
import {
    isAny,
    isArbitraryImage,
    isArbitraryLength,
    isArbitraryNumber,
    isArbitraryPosition,
    isArbitraryShadow,
    isArbitrarySize,
    isArbitraryValue,
    isInteger,
    isLength,
    isNumber,
    isPercent,
    isTshirtSize,
} from './validators'

export const getDefaultConfig = () => {
    const colors = fromTheme(&apos;colors&apos;)
    const spacing = fromTheme(&apos;spacing&apos;)
    const blur = fromTheme(&apos;blur&apos;)
    const brightness = fromTheme(&apos;brightness&apos;)
    const borderColor = fromTheme(&apos;borderColor&apos;)
    const borderRadius = fromTheme(&apos;borderRadius&apos;)
    const borderSpacing = fromTheme(&apos;borderSpacing&apos;)
    const borderWidth = fromTheme(&apos;borderWidth&apos;)
    const contrast = fromTheme(&apos;contrast&apos;)
    const grayscale = fromTheme(&apos;grayscale&apos;)
    const hueRotate = fromTheme(&apos;hueRotate&apos;)
    const invert = fromTheme(&apos;invert&apos;)
    const gap = fromTheme(&apos;gap&apos;)
    const gradientColorStops = fromTheme(&apos;gradientColorStops&apos;)
    const gradientColorStopPositions = fromTheme(&apos;gradientColorStopPositions&apos;)
    const inset = fromTheme(&apos;inset&apos;)
    const margin = fromTheme(&apos;margin&apos;)
    const opacity = fromTheme(&apos;opacity&apos;)
    const padding = fromTheme(&apos;padding&apos;)
    const saturate = fromTheme(&apos;saturate&apos;)
    const scale = fromTheme(&apos;scale&apos;)
    const sepia = fromTheme(&apos;sepia&apos;)
    const skew = fromTheme(&apos;skew&apos;)
    const space = fromTheme(&apos;space&apos;)
    const translate = fromTheme(&apos;translate&apos;)

    const getOverscroll = () => [&apos;auto&apos;, &apos;contain&apos;, &apos;none&apos;] as const
    const getOverflow = () => [&apos;auto&apos;, &apos;hidden&apos;, &apos;clip&apos;, &apos;visible&apos;, &apos;scroll&apos;] as const
    const getSpacingWithAutoAndArbitrary = () => [&apos;auto&apos;, isArbitraryValue, spacing] as const
    const getSpacingWithArbitrary = () => [isArbitraryValue, spacing] as const
    const getLengthWithEmptyAndArbitrary = () => [&apos;&apos;, isLength, isArbitraryLength] as const
    const getNumberWithAutoAndArbitrary = () => [&apos;auto&apos;, isNumber, isArbitraryValue] as const
    const getPositions = () =>
        [
            &apos;bottom&apos;,
            &apos;center&apos;,
            &apos;left&apos;,
            &apos;left-bottom&apos;,
            &apos;left-top&apos;,
            &apos;right&apos;,
            &apos;right-bottom&apos;,
            &apos;right-top&apos;,
            &apos;top&apos;,
        ] as const
    const getLineStyles = () => [&apos;solid&apos;, &apos;dashed&apos;, &apos;dotted&apos;, &apos;double&apos;, &apos;none&apos;] as const
    const getBlendModes = () =>
        [
            &apos;normal&apos;,
            &apos;multiply&apos;,
            &apos;screen&apos;,
            &apos;overlay&apos;,
            &apos;darken&apos;,
            &apos;lighten&apos;,
            &apos;color-dodge&apos;,
            &apos;color-burn&apos;,
            &apos;hard-light&apos;,
            &apos;soft-light&apos;,
            &apos;difference&apos;,
            &apos;exclusion&apos;,
            &apos;hue&apos;,
            &apos;saturation&apos;,
            &apos;color&apos;,
            &apos;luminosity&apos;,
        ] as const
    const getAlign = () =>
        [&apos;start&apos;, &apos;end&apos;, &apos;center&apos;, &apos;between&apos;, &apos;around&apos;, &apos;evenly&apos;, &apos;stretch&apos;] as const
    const getZeroAndEmpty = () => [&apos;&apos;, &apos;0&apos;, isArbitraryValue] as const
    const getBreaks = () =>
        [&apos;auto&apos;, &apos;avoid&apos;, &apos;all&apos;, &apos;avoid-page&apos;, &apos;page&apos;, &apos;left&apos;, &apos;right&apos;, &apos;column&apos;] as const
    const getNumberAndArbitrary = () => [isNumber, isArbitraryValue]

    return {
        cacheSize: 500,
        separator: &apos;:&apos;,
        theme: {
            colors: [isAny],
            spacing: [isLength, isArbitraryLength],
            blur: [&apos;none&apos;, &apos;&apos;, isTshirtSize, isArbitraryValue],
            brightness: getNumberAndArbitrary(),
            borderColor: [colors],
            borderRadius: [&apos;none&apos;, &apos;&apos;, &apos;full&apos;, isTshirtSize, isArbitraryValue],
            borderSpacing: getSpacingWithArbitrary(),
            borderWidth: getLengthWithEmptyAndArbitrary(),
            contrast: getNumberAndArbitrary(),
            grayscale: getZeroAndEmpty(),
            hueRotate: getNumberAndArbitrary(),
            invert: getZeroAndEmpty(),
            gap: getSpacingWithArbitrary(),
            gradientColorStops: [colors],
            gradientColorStopPositions: [isPercent, isArbitraryLength],
            inset: getSpacingWithAutoAndArbitrary(),
            margin: getSpacingWithAutoAndArbitrary(),
            opacity: getNumberAndArbitrary(),
            padding: getSpacingWithArbitrary(),
            saturate: getNumberAndArbitrary(),
            scale: getNumberAndArbitrary(),
            sepia: getZeroAndEmpty(),
            skew: getNumberAndArbitrary(),
            space: getSpacingWithArbitrary(),
            translate: getSpacingWithArbitrary(),
        },
        classGroups: {
            // Layout
            /**
             * Aspect Ratio
             * @see https://tailwindcss.com/docs/aspect-ratio
             */
            aspect: [{ aspect: [&apos;auto&apos;, &apos;square&apos;, &apos;video&apos;, isArbitraryValue] }],
            /**
             * Container
             * @see https://tailwindcss.com/docs/container
             */
            container: [&apos;container&apos;],
            /**
             * Columns
             * @see https://tailwindcss.com/docs/columns
             */
            columns: [{ columns: [isTshirtSize] }],
            /**
             * Break After
             * @see https://tailwindcss.com/docs/break-after
             */
            &apos;break-after&apos;: [{ &apos;break-after&apos;: getBreaks() }],
            /**
             * Break Before
             * @see https://tailwindcss.com/docs/break-before
             */
            &apos;break-before&apos;: [{ &apos;break-before&apos;: getBreaks() }],
            /**
             * Break Inside
             * @see https://tailwindcss.com/docs/break-inside
             */
            &apos;break-inside&apos;: [{ &apos;break-inside&apos;: [&apos;auto&apos;, &apos;avoid&apos;, &apos;avoid-page&apos;, &apos;avoid-column&apos;] }],
            /**
             * Box Decoration Break
             * @see https://tailwindcss.com/docs/box-decoration-break
             */
            &apos;box-decoration&apos;: [{ &apos;box-decoration&apos;: [&apos;slice&apos;, &apos;clone&apos;] }],
            /**
             * Box Sizing
             * @see https://tailwindcss.com/docs/box-sizing
             */
            box: [{ box: [&apos;border&apos;, &apos;content&apos;] }],
            /**
             * Display
             * @see https://tailwindcss.com/docs/display
             */
            display: [
                &apos;block&apos;,
                &apos;inline-block&apos;,
                &apos;inline&apos;,
                &apos;flex&apos;,
                &apos;inline-flex&apos;,
                &apos;table&apos;,
                &apos;inline-table&apos;,
                &apos;table-caption&apos;,
                &apos;table-cell&apos;,
                &apos;table-column&apos;,
                &apos;table-column-group&apos;,
                &apos;table-footer-group&apos;,
                &apos;table-header-group&apos;,
                &apos;table-row-group&apos;,
                &apos;table-row&apos;,
                &apos;flow-root&apos;,
                &apos;grid&apos;,
                &apos;inline-grid&apos;,
                &apos;contents&apos;,
                &apos;list-item&apos;,
                &apos;hidden&apos;,
            ],
            /**
             * Floats
             * @see https://tailwindcss.com/docs/float
             */
            float: [{ float: [&apos;right&apos;, &apos;left&apos;, &apos;none&apos;, &apos;start&apos;, &apos;end&apos;] }],
            /**
             * Clear
             * @see https://tailwindcss.com/docs/clear
             */
            clear: [{ clear: [&apos;left&apos;, &apos;right&apos;, &apos;both&apos;, &apos;none&apos;, &apos;start&apos;, &apos;end&apos;] }],
            /**
             * Isolation
             * @see https://tailwindcss.com/docs/isolation
             */
            isolation: [&apos;isolate&apos;, &apos;isolation-auto&apos;],
            /**
             * Object Fit
             * @see https://tailwindcss.com/docs/object-fit
             */
            &apos;object-fit&apos;: [{ object: [&apos;contain&apos;, &apos;cover&apos;, &apos;fill&apos;, &apos;none&apos;, &apos;scale-down&apos;] }],
            /**
             * Object Position
             * @see https://tailwindcss.com/docs/object-position
             */
            &apos;object-position&apos;: [{ object: [...getPositions(), isArbitraryValue] }],
            /**
             * Overflow
             * @see https://tailwindcss.com/docs/overflow
             */
            overflow: [{ overflow: getOverflow() }],
            /**
             * Overflow X
             * @see https://tailwindcss.com/docs/overflow
             */
            &apos;overflow-x&apos;: [{ &apos;overflow-x&apos;: getOverflow() }],
            /**
             * Overflow Y
             * @see https://tailwindcss.com/docs/overflow
             */
            &apos;overflow-y&apos;: [{ &apos;overflow-y&apos;: getOverflow() }],
            /**
             * Overscroll Behavior
             * @see https://tailwindcss.com/docs/overscroll-behavior
             */
            overscroll: [{ overscroll: getOverscroll() }],
            /**
             * Overscroll Behavior X
             * @see https://tailwindcss.com/docs/overscroll-behavior
             */
            &apos;overscroll-x&apos;: [{ &apos;overscroll-x&apos;: getOverscroll() }],
            /**
             * Overscroll Behavior Y
             * @see https://tailwindcss.com/docs/overscroll-behavior
             */
            &apos;overscroll-y&apos;: [{ &apos;overscroll-y&apos;: getOverscroll() }],
            /**
             * Position
             * @see https://tailwindcss.com/docs/position
             */
            position: [&apos;static&apos;, &apos;fixed&apos;, &apos;absolute&apos;, &apos;relative&apos;, &apos;sticky&apos;],
            /**
             * Top / Right / Bottom / Left
             * @see https://tailwindcss.com/docs/top-right-bottom-left
             */
            inset: [{ inset: [inset] }],
            /**
             * Right / Left
             * @see https://tailwindcss.com/docs/top-right-bottom-left
             */
            &apos;inset-x&apos;: [{ &apos;inset-x&apos;: [inset] }],
            /**
             * Top / Bottom
             * @see https://tailwindcss.com/docs/top-right-bottom-left
             */
            &apos;inset-y&apos;: [{ &apos;inset-y&apos;: [inset] }],
            /**
             * Start
             * @see https://tailwindcss.com/docs/top-right-bottom-left
             */
            start: [{ start: [inset] }],
            /**
             * End
             * @see https://tailwindcss.com/docs/top-right-bottom-left
             */
            end: [{ end: [inset] }],
            /**
             * Top
             * @see https://tailwindcss.com/docs/top-right-bottom-left
             */
            top: [{ top: [inset] }],
            /**
             * Right
             * @see https://tailwindcss.com/docs/top-right-bottom-left
             */
            right: [{ right: [inset] }],
            /**
             * Bottom
             * @see https://tailwindcss.com/docs/top-right-bottom-left
             */
            bottom: [{ bottom: [inset] }],
            /**
             * Left
             * @see https://tailwindcss.com/docs/top-right-bottom-left
             */
            left: [{ left: [inset] }],
            /**
             * Visibility
             * @see https://tailwindcss.com/docs/visibility
             */
            visibility: [&apos;visible&apos;, &apos;invisible&apos;, &apos;collapse&apos;],
            /**
             * Z-Index
             * @see https://tailwindcss.com/docs/z-index
             */
            z: [{ z: [&apos;auto&apos;, isInteger, isArbitraryValue] }],
            // Flexbox and Grid
            /**
             * Flex Basis
             * @see https://tailwindcss.com/docs/flex-basis
             */
            basis: [{ basis: getSpacingWithAutoAndArbitrary() }],
            /**
             * Flex Direction
             * @see https://tailwindcss.com/docs/flex-direction
             */
            &apos;flex-direction&apos;: [{ flex: [&apos;row&apos;, &apos;row-reverse&apos;, &apos;col&apos;, &apos;col-reverse&apos;] }],
            /**
             * Flex Wrap
             * @see https://tailwindcss.com/docs/flex-wrap
             */
            &apos;flex-wrap&apos;: [{ flex: [&apos;wrap&apos;, &apos;wrap-reverse&apos;, &apos;nowrap&apos;] }],
            /**
             * Flex
             * @see https://tailwindcss.com/docs/flex
             */
            flex: [{ flex: [&apos;1&apos;, &apos;auto&apos;, &apos;initial&apos;, &apos;none&apos;, isArbitraryValue] }],
            /**
             * Flex Grow
             * @see https://tailwindcss.com/docs/flex-grow
             */
            grow: [{ grow: getZeroAndEmpty() }],
            /**
             * Flex Shrink
             * @see https://tailwindcss.com/docs/flex-shrink
             */
            shrink: [{ shrink: getZeroAndEmpty() }],
            /**
             * Order
             * @see https://tailwindcss.com/docs/order
             */
            order: [{ order: [&apos;first&apos;, &apos;last&apos;, &apos;none&apos;, isInteger, isArbitraryValue] }],
            /**
             * Grid Template Columns
             * @see https://tailwindcss.com/docs/grid-template-columns
             */
            &apos;grid-cols&apos;: [{ &apos;grid-cols&apos;: [isAny] }],
            /**
             * Grid Column Start / End
             * @see https://tailwindcss.com/docs/grid-column
             */
            &apos;col-start-end&apos;: [
                {
                    col: [
                        &apos;auto&apos;,
                        { span: [&apos;full&apos;, isInteger, isArbitraryValue] },
                        isArbitraryValue,
                    ],
                },
            ],
            /**
             * Grid Column Start
             * @see https://tailwindcss.com/docs/grid-column
             */
            &apos;col-start&apos;: [{ &apos;col-start&apos;: getNumberWithAutoAndArbitrary() }],
            /**
             * Grid Column End
             * @see https://tailwindcss.com/docs/grid-column
             */
            &apos;col-end&apos;: [{ &apos;col-end&apos;: getNumberWithAutoAndArbitrary() }],
            /**
             * Grid Template Rows
             * @see https://tailwindcss.com/docs/grid-template-rows
             */
            &apos;grid-rows&apos;: [{ &apos;grid-rows&apos;: [isAny] }],
            /**
             * Grid Row Start / End
             * @see https://tailwindcss.com/docs/grid-row
             */
            &apos;row-start-end&apos;: [
                { row: [&apos;auto&apos;, { span: [isInteger, isArbitraryValue] }, isArbitraryValue] },
            ],
            /**
             * Grid Row Start
             * @see https://tailwindcss.com/docs/grid-row
             */
            &apos;row-start&apos;: [{ &apos;row-start&apos;: getNumberWithAutoAndArbitrary() }],
            /**
             * Grid Row End
             * @see https://tailwindcss.com/docs/grid-row
             */
            &apos;row-end&apos;: [{ &apos;row-end&apos;: getNumberWithAutoAndArbitrary() }],
            /**
             * Grid Auto Flow
             * @see https://tailwindcss.com/docs/grid-auto-flow
             */
            &apos;grid-flow&apos;: [{ &apos;grid-flow&apos;: [&apos;row&apos;, &apos;col&apos;, &apos;dense&apos;, &apos;row-dense&apos;, &apos;col-dense&apos;] }],
            /**
             * Grid Auto Columns
             * @see https://tailwindcss.com/docs/grid-auto-columns
             */
            &apos;auto-cols&apos;: [{ &apos;auto-cols&apos;: [&apos;auto&apos;, &apos;min&apos;, &apos;max&apos;, &apos;fr&apos;, isArbitraryValue] }],
            /**
             * Grid Auto Rows
             * @see https://tailwindcss.com/docs/grid-auto-rows
             */
            &apos;auto-rows&apos;: [{ &apos;auto-rows&apos;: [&apos;auto&apos;, &apos;min&apos;, &apos;max&apos;, &apos;fr&apos;, isArbitraryValue] }],
            /**
             * Gap
             * @see https://tailwindcss.com/docs/gap
             */
            gap: [{ gap: [gap] }],
            /**
             * Gap X
             * @see https://tailwindcss.com/docs/gap
             */
            &apos;gap-x&apos;: [{ &apos;gap-x&apos;: [gap] }],
            /**
             * Gap Y
             * @see https://tailwindcss.com/docs/gap
             */
            &apos;gap-y&apos;: [{ &apos;gap-y&apos;: [gap] }],
            /**
             * Justify Content
             * @see https://tailwindcss.com/docs/justify-content
             */
            &apos;justify-content&apos;: [{ justify: [&apos;normal&apos;, ...getAlign()] }],
            /**
             * Justify Items
             * @see https://tailwindcss.com/docs/justify-items
             */
            &apos;justify-items&apos;: [{ &apos;justify-items&apos;: [&apos;start&apos;, &apos;end&apos;, &apos;center&apos;, &apos;stretch&apos;] }],
            /**
             * Justify Self
             * @see https://tailwindcss.com/docs/justify-self
             */
            &apos;justify-self&apos;: [{ &apos;justify-self&apos;: [&apos;auto&apos;, &apos;start&apos;, &apos;end&apos;, &apos;center&apos;, &apos;stretch&apos;] }],
            /**
             * Align Content
             * @see https://tailwindcss.com/docs/align-content
             */
            &apos;align-content&apos;: [{ content: [&apos;normal&apos;, ...getAlign(), &apos;baseline&apos;] }],
            /**
             * Align Items
             * @see https://tailwindcss.com/docs/align-items
             */
            &apos;align-items&apos;: [{ items: [&apos;start&apos;, &apos;end&apos;, &apos;center&apos;, &apos;baseline&apos;, &apos;stretch&apos;] }],
            /**
             * Align Self
             * @see https://tailwindcss.com/docs/align-self
             */
            &apos;align-self&apos;: [{ self: [&apos;auto&apos;, &apos;start&apos;, &apos;end&apos;, &apos;center&apos;, &apos;stretch&apos;, &apos;baseline&apos;] }],
            /**
             * Place Content
             * @see https://tailwindcss.com/docs/place-content
             */
            &apos;place-content&apos;: [{ &apos;place-content&apos;: [...getAlign(), &apos;baseline&apos;] }],
            /**
             * Place Items
             * @see https://tailwindcss.com/docs/place-items
             */
            &apos;place-items&apos;: [{ &apos;place-items&apos;: [&apos;start&apos;, &apos;end&apos;, &apos;center&apos;, &apos;baseline&apos;, &apos;stretch&apos;] }],
            /**
             * Place Self
             * @see https://tailwindcss.com/docs/place-self
             */
            &apos;place-self&apos;: [{ &apos;place-self&apos;: [&apos;auto&apos;, &apos;start&apos;, &apos;end&apos;, &apos;center&apos;, &apos;stretch&apos;] }],
            // Spacing
            /**
             * Padding
             * @see https://tailwindcss.com/docs/padding
             */
            p: [{ p: [padding] }],
            /**
             * Padding X
             * @see https://tailwindcss.com/docs/padding
             */
            px: [{ px: [padding] }],
            /**
             * Padding Y
             * @see https://tailwindcss.com/docs/padding
             */
            py: [{ py: [padding] }],
            /**
             * Padding Start
             * @see https://tailwindcss.com/docs/padding
             */
            ps: [{ ps: [padding] }],
            /**
             * Padding End
             * @see https://tailwindcss.com/docs/padding
             */
            pe: [{ pe: [padding] }],
            /**
             * Padding Top
             * @see https://tailwindcss.com/docs/padding
             */
            pt: [{ pt: [padding] }],
            /**
             * Padding Right
             * @see https://tailwindcss.com/docs/padding
             */
            pr: [{ pr: [padding] }],
            /**
             * Padding Bottom
             * @see https://tailwindcss.com/docs/padding
             */
            pb: [{ pb: [padding] }],
            /**
             * Padding Left
             * @see https://tailwindcss.com/docs/padding
             */
            pl: [{ pl: [padding] }],
            /**
             * Margin
             * @see https://tailwindcss.com/docs/margin
             */
            m: [{ m: [margin] }],
            /**
             * Margin X
             * @see https://tailwindcss.com/docs/margin
             */
            mx: [{ mx: [margin] }],
            /**
             * Margin Y
             * @see https://tailwindcss.com/docs/margin
             */
            my: [{ my: [margin] }],
            /**
             * Margin Start
             * @see https://tailwindcss.com/docs/margin
             */
            ms: [{ ms: [margin] }],
            /**
             * Margin End
             * @see https://tailwindcss.com/docs/margin
             */
            me: [{ me: [margin] }],
            /**
             * Margin Top
             * @see https://tailwindcss.com/docs/margin
             */
            mt: [{ mt: [margin] }],
            /**
             * Margin Right
             * @see https://tailwindcss.com/docs/margin
             */
            mr: [{ mr: [margin] }],
            /**
             * Margin Bottom
             * @see https://tailwindcss.com/docs/margin
             */
            mb: [{ mb: [margin] }],
            /**
             * Margin Left
             * @see https://tailwindcss.com/docs/margin
             */
            ml: [{ ml: [margin] }],
            /**
             * Space Between X
             * @see https://tailwindcss.com/docs/space
             */
            &apos;space-x&apos;: [{ &apos;space-x&apos;: [space] }],
            /**
             * Space Between X Reverse
             * @see https://tailwindcss.com/docs/space
             */
            &apos;space-x-reverse&apos;: [&apos;space-x-reverse&apos;],
            /**
             * Space Between Y
             * @see https://tailwindcss.com/docs/space
             */
            &apos;space-y&apos;: [{ &apos;space-y&apos;: [space] }],
            /**
             * Space Between Y Reverse
             * @see https://tailwindcss.com/docs/space
             */
            &apos;space-y-reverse&apos;: [&apos;space-y-reverse&apos;],
            // Sizing
            /**
             * Width
             * @see https://tailwindcss.com/docs/width
             */
            w: [
                {
                    w: [
                        &apos;auto&apos;,
                        &apos;min&apos;,
                        &apos;max&apos;,
                        &apos;fit&apos;,
                        &apos;svw&apos;,
                        &apos;lvw&apos;,
                        &apos;dvw&apos;,
                        isArbitraryValue,
                        spacing,
                    ],
                },
            ],
            /**
             * Min-Width
             * @see https://tailwindcss.com/docs/min-width
             */
            &apos;min-w&apos;: [{ &apos;min-w&apos;: [isArbitraryValue, spacing, &apos;min&apos;, &apos;max&apos;, &apos;fit&apos;] }],
            /**
             * Max-Width
             * @see https://tailwindcss.com/docs/max-width
             */
            &apos;max-w&apos;: [
                {
                    &apos;max-w&apos;: [
                        isArbitraryValue,
                        spacing,
                        &apos;none&apos;,
                        &apos;full&apos;,
                        &apos;min&apos;,
                        &apos;max&apos;,
                        &apos;fit&apos;,
                        &apos;prose&apos;,
                        { screen: [isTshirtSize] },
                        isTshirtSize,
                    ],
                },
            ],
            /**
             * Height
             * @see https://tailwindcss.com/docs/height
             */
            h: [
                {
                    h: [
                        isArbitraryValue,
                        spacing,
                        &apos;auto&apos;,
                        &apos;min&apos;,
                        &apos;max&apos;,
                        &apos;fit&apos;,
                        &apos;svh&apos;,
                        &apos;lvh&apos;,
                        &apos;dvh&apos;,
                    ],
                },
            ],
            /**
             * Min-Height
             * @see https://tailwindcss.com/docs/min-height
             */
            &apos;min-h&apos;: [
                { &apos;min-h&apos;: [isArbitraryValue, spacing, &apos;min&apos;, &apos;max&apos;, &apos;fit&apos;, &apos;svh&apos;, &apos;lvh&apos;, &apos;dvh&apos;] },
            ],
            /**
             * Max-Height
             * @see https://tailwindcss.com/docs/max-height
             */
            &apos;max-h&apos;: [
                { &apos;max-h&apos;: [isArbitraryValue, spacing, &apos;min&apos;, &apos;max&apos;, &apos;fit&apos;, &apos;svh&apos;, &apos;lvh&apos;, &apos;dvh&apos;] },
            ],
            /**
             * Size
             * @see https://tailwindcss.com/docs/size
             */
            size: [{ size: [isArbitraryValue, spacing, &apos;auto&apos;, &apos;min&apos;, &apos;max&apos;, &apos;fit&apos;] }],
            // Typography
            /**
             * Font Size
             * @see https://tailwindcss.com/docs/font-size
             */
            &apos;font-size&apos;: [{ text: [&apos;base&apos;, isTshirtSize, isArbitraryLength] }],
            /**
             * Font Smoothing
             * @see https://tailwindcss.com/docs/font-smoothing
             */
            &apos;font-smoothing&apos;: [&apos;antialiased&apos;, &apos;subpixel-antialiased&apos;],
            /**
             * Font Style
             * @see https://tailwindcss.com/docs/font-style
             */
            &apos;font-style&apos;: [&apos;italic&apos;, &apos;not-italic&apos;],
            /**
             * Font Weight
             * @see https://tailwindcss.com/docs/font-weight
             */
            &apos;font-weight&apos;: [
                {
                    font: [
                        &apos;thin&apos;,
                        &apos;extralight&apos;,
                        &apos;light&apos;,
                        &apos;normal&apos;,
                        &apos;medium&apos;,
                        &apos;semibold&apos;,
                        &apos;bold&apos;,
                        &apos;extrabold&apos;,
                        &apos;black&apos;,
                        isArbitraryNumber,
                    ],
                },
            ],
            /**
             * Font Family
             * @see https://tailwindcss.com/docs/font-family
             */
            &apos;font-family&apos;: [{ font: [isAny] }],
            /**
             * Font Variant Numeric
             * @see https://tailwindcss.com/docs/font-variant-numeric
             */
            &apos;fvn-normal&apos;: [&apos;normal-nums&apos;],
            /**
             * Font Variant Numeric
             * @see https://tailwindcss.com/docs/font-variant-numeric
             */
            &apos;fvn-ordinal&apos;: [&apos;ordinal&apos;],
            /**
             * Font Variant Numeric
             * @see https://tailwindcss.com/docs/font-variant-numeric
             */
            &apos;fvn-slashed-zero&apos;: [&apos;slashed-zero&apos;],
            /**
             * Font Variant Numeric
             * @see https://tailwindcss.com/docs/font-variant-numeric
             */
            &apos;fvn-figure&apos;: [&apos;lining-nums&apos;, &apos;oldstyle-nums&apos;],
            /**
             * Font Variant Numeric
             * @see https://tailwindcss.com/docs/font-variant-numeric
             */
            &apos;fvn-spacing&apos;: [&apos;proportional-nums&apos;, &apos;tabular-nums&apos;],
            /**
             * Font Variant Numeric
             * @see https://tailwindcss.com/docs/font-variant-numeric
             */
            &apos;fvn-fraction&apos;: [&apos;diagonal-fractions&apos;, &apos;stacked-fractions&apos;],
            /**
             * Letter Spacing
             * @see https://tailwindcss.com/docs/letter-spacing
             */
            tracking: [
                {
                    tracking: [
                        &apos;tighter&apos;,
                        &apos;tight&apos;,
                        &apos;normal&apos;,
                        &apos;wide&apos;,
                        &apos;wider&apos;,
                        &apos;widest&apos;,
                        isArbitraryValue,
                    ],
                },
            ],
            /**
             * Line Clamp
             * @see https://tailwindcss.com/docs/line-clamp
             */
            &apos;line-clamp&apos;: [{ &apos;line-clamp&apos;: [&apos;none&apos;, isNumber, isArbitraryNumber] }],
            /**
             * Line Height
             * @see https://tailwindcss.com/docs/line-height
             */
            leading: [
                {
                    leading: [
                        &apos;none&apos;,
                        &apos;tight&apos;,
                        &apos;snug&apos;,
                        &apos;normal&apos;,
                        &apos;relaxed&apos;,
                        &apos;loose&apos;,
                        isLength,
                        isArbitraryValue,
                    ],
                },
            ],
            /**
             * List Style Image
             * @see https://tailwindcss.com/docs/list-style-image
             */
            &apos;list-image&apos;: [{ &apos;list-image&apos;: [&apos;none&apos;, isArbitraryValue] }],
            /**
             * List Style Type
             * @see https://tailwindcss.com/docs/list-style-type
             */
            &apos;list-style-type&apos;: [{ list: [&apos;none&apos;, &apos;disc&apos;, &apos;decimal&apos;, isArbitraryValue] }],
            /**
             * List Style Position
             * @see https://tailwindcss.com/docs/list-style-position
             */
            &apos;list-style-position&apos;: [{ list: [&apos;inside&apos;, &apos;outside&apos;] }],
            /**
             * Placeholder Color
             * @deprecated since Tailwind CSS v3.0.0
             * @see https://tailwindcss.com/docs/placeholder-color
             */
            &apos;placeholder-color&apos;: [{ placeholder: [colors] }],
            /**
             * Placeholder Opacity
             * @see https://tailwindcss.com/docs/placeholder-opacity
             */
            &apos;placeholder-opacity&apos;: [{ &apos;placeholder-opacity&apos;: [opacity] }],
            /**
             * Text Alignment
             * @see https://tailwindcss.com/docs/text-align
             */
            &apos;text-alignment&apos;: [{ text: [&apos;left&apos;, &apos;center&apos;, &apos;right&apos;, &apos;justify&apos;, &apos;start&apos;, &apos;end&apos;] }],
            /**
             * Text Color
             * @see https://tailwindcss.com/docs/text-color
             */
            &apos;text-color&apos;: [{ text: [colors] }],
            /**
             * Text Opacity
             * @see https://tailwindcss.com/docs/text-opacity
             */
            &apos;text-opacity&apos;: [{ &apos;text-opacity&apos;: [opacity] }],
            /**
             * Text Decoration
             * @see https://tailwindcss.com/docs/text-decoration
             */
            &apos;text-decoration&apos;: [&apos;underline&apos;, &apos;overline&apos;, &apos;line-through&apos;, &apos;no-underline&apos;],
            /**
             * Text Decoration Style
             * @see https://tailwindcss.com/docs/text-decoration-style
             */
            &apos;text-decoration-style&apos;: [{ decoration: [...getLineStyles(), &apos;wavy&apos;] }],
            /**
             * Text Decoration Thickness
             * @see https://tailwindcss.com/docs/text-decoration-thickness
             */
            &apos;text-decoration-thickness&apos;: [
                { decoration: [&apos;auto&apos;, &apos;from-font&apos;, isLength, isArbitraryLength] },
            ],
            /**
             * Text Underline Offset
             * @see https://tailwindcss.com/docs/text-underline-offset
             */
            &apos;underline-offset&apos;: [{ &apos;underline-offset&apos;: [&apos;auto&apos;, isLength, isArbitraryValue] }],
            /**
             * Text Decoration Color
             * @see https://tailwindcss.com/docs/text-decoration-color
             */
            &apos;text-decoration-color&apos;: [{ decoration: [colors] }],
            /**
             * Text Transform
             * @see https://tailwindcss.com/docs/text-transform
             */
            &apos;text-transform&apos;: [&apos;uppercase&apos;, &apos;lowercase&apos;, &apos;capitalize&apos;, &apos;normal-case&apos;],
            /**
             * Text Overflow
             * @see https://tailwindcss.com/docs/text-overflow
             */
            &apos;text-overflow&apos;: [&apos;truncate&apos;, &apos;text-ellipsis&apos;, &apos;text-clip&apos;],
            /**
             * Text Wrap
             * @see https://tailwindcss.com/docs/text-wrap
             */
            &apos;text-wrap&apos;: [{ text: [&apos;wrap&apos;, &apos;nowrap&apos;, &apos;balance&apos;, &apos;pretty&apos;] }],
            /**
             * Text Indent
             * @see https://tailwindcss.com/docs/text-indent
             */
            indent: [{ indent: getSpacingWithArbitrary() }],
            /**
             * Vertical Alignment
             * @see https://tailwindcss.com/docs/vertical-align
             */
            &apos;vertical-align&apos;: [
                {
                    align: [
                        &apos;baseline&apos;,
                        &apos;top&apos;,
                        &apos;middle&apos;,
                        &apos;bottom&apos;,
                        &apos;text-top&apos;,
                        &apos;text-bottom&apos;,
                        &apos;sub&apos;,
                        &apos;super&apos;,
                        isArbitraryValue,
                    ],
                },
            ],
            /**
             * Whitespace
             * @see https://tailwindcss.com/docs/whitespace
             */
            whitespace: [
                { whitespace: [&apos;normal&apos;, &apos;nowrap&apos;, &apos;pre&apos;, &apos;pre-line&apos;, &apos;pre-wrap&apos;, &apos;break-spaces&apos;] },
            ],
            /**
             * Word Break
             * @see https://tailwindcss.com/docs/word-break
             */
            break: [{ break: [&apos;normal&apos;, &apos;words&apos;, &apos;all&apos;, &apos;keep&apos;] }],
            /**
             * Hyphens
             * @see https://tailwindcss.com/docs/hyphens
             */
            hyphens: [{ hyphens: [&apos;none&apos;, &apos;manual&apos;, &apos;auto&apos;] }],
            /**
             * Content
             * @see https://tailwindcss.com/docs/content
             */
            content: [{ content: [&apos;none&apos;, isArbitraryValue] }],
            // Backgrounds
            /**
             * Background Attachment
             * @see https://tailwindcss.com/docs/background-attachment
             */
            &apos;bg-attachment&apos;: [{ bg: [&apos;fixed&apos;, &apos;local&apos;, &apos;scroll&apos;] }],
            /**
             * Background Clip
             * @see https://tailwindcss.com/docs/background-clip
             */
            &apos;bg-clip&apos;: [{ &apos;bg-clip&apos;: [&apos;border&apos;, &apos;padding&apos;, &apos;content&apos;, &apos;text&apos;] }],
            /**
             * Background Opacity
             * @deprecated since Tailwind CSS v3.0.0
             * @see https://tailwindcss.com/docs/background-opacity
             */
            &apos;bg-opacity&apos;: [{ &apos;bg-opacity&apos;: [opacity] }],
            /**
             * Background Origin
             * @see https://tailwindcss.com/docs/background-origin
             */
            &apos;bg-origin&apos;: [{ &apos;bg-origin&apos;: [&apos;border&apos;, &apos;padding&apos;, &apos;content&apos;] }],
            /**
             * Background Position
             * @see https://tailwindcss.com/docs/background-position
             */
            &apos;bg-position&apos;: [{ bg: [...getPositions(), isArbitraryPosition] }],
            /**
             * Background Repeat
             * @see https://tailwindcss.com/docs/background-repeat
             */
            &apos;bg-repeat&apos;: [{ bg: [&apos;no-repeat&apos;, { repeat: [&apos;&apos;, &apos;x&apos;, &apos;y&apos;, &apos;round&apos;, &apos;space&apos;] }] }],
            /**
             * Background Size
             * @see https://tailwindcss.com/docs/background-size
             */
            &apos;bg-size&apos;: [{ bg: [&apos;auto&apos;, &apos;cover&apos;, &apos;contain&apos;, isArbitrarySize] }],
            /**
             * Background Image
             * @see https://tailwindcss.com/docs/background-image
             */
            &apos;bg-image&apos;: [
                {
                    bg: [
                        &apos;none&apos;,
                        { &apos;gradient-to&apos;: [&apos;t&apos;, &apos;tr&apos;, &apos;r&apos;, &apos;br&apos;, &apos;b&apos;, &apos;bl&apos;, &apos;l&apos;, &apos;tl&apos;] },
                        isArbitraryImage,
                    ],
                },
            ],
            /**
             * Background Color
             * @see https://tailwindcss.com/docs/background-color
             */
            &apos;bg-color&apos;: [{ bg: [colors] }],
            /**
             * Gradient Color Stops From Position
             * @see https://tailwindcss.com/docs/gradient-color-stops
             */
            &apos;gradient-from-pos&apos;: [{ from: [gradientColorStopPositions] }],
            /**
             * Gradient Color Stops Via Position
             * @see https://tailwindcss.com/docs/gradient-color-stops
             */
            &apos;gradient-via-pos&apos;: [{ via: [gradientColorStopPositions] }],
            /**
             * Gradient Color Stops To Position
             * @see https://tailwindcss.com/docs/gradient-color-stops
             */
            &apos;gradient-to-pos&apos;: [{ to: [gradientColorStopPositions] }],
            /**
             * Gradient Color Stops From
             * @see https://tailwindcss.com/docs/gradient-color-stops
             */
            &apos;gradient-from&apos;: [{ from: [gradientColorStops] }],
            /**
             * Gradient Color Stops Via
             * @see https://tailwindcss.com/docs/gradient-color-stops
             */
            &apos;gradient-via&apos;: [{ via: [gradientColorStops] }],
            /**
             * Gradient Color Stops To
             * @see https://tailwindcss.com/docs/gradient-color-stops
             */
            &apos;gradient-to&apos;: [{ to: [gradientColorStops] }],
            // Borders
            /**
             * Border Radius
             * @see https://tailwindcss.com/docs/border-radius
             */
            rounded: [{ rounded: [borderRadius] }],
            /**
             * Border Radius Start
             * @see https://tailwindcss.com/docs/border-radius
             */
            &apos;rounded-s&apos;: [{ &apos;rounded-s&apos;: [borderRadius] }],
            /**
             * Border Radius End
             * @see https://tailwindcss.com/docs/border-radius
             */
            &apos;rounded-e&apos;: [{ &apos;rounded-e&apos;: [borderRadius] }],
            /**
             * Border Radius Top
             * @see https://tailwindcss.com/docs/border-radius
             */
            &apos;rounded-t&apos;: [{ &apos;rounded-t&apos;: [borderRadius] }],
            /**
             * Border Radius Right
             * @see https://tailwindcss.com/docs/border-radius
             */
            &apos;rounded-r&apos;: [{ &apos;rounded-r&apos;: [borderRadius] }],
            /**
             * Border Radius Bottom
             * @see https://tailwindcss.com/docs/border-radius
             */
            &apos;rounded-b&apos;: [{ &apos;rounded-b&apos;: [borderRadius] }],
            /**
             * Border Radius Left
             * @see https://tailwindcss.com/docs/border-radius
             */
            &apos;rounded-l&apos;: [{ &apos;rounded-l&apos;: [borderRadius] }],
            /**
             * Border Radius Start Start
             * @see https://tailwindcss.com/docs/border-radius
             */
            &apos;rounded-ss&apos;: [{ &apos;rounded-ss&apos;: [borderRadius] }],
            /**
             * Border Radius Start End
             * @see https://tailwindcss.com/docs/border-radius
             */
            &apos;rounded-se&apos;: [{ &apos;rounded-se&apos;: [borderRadius] }],
            /**
             * Border Radius End End
             * @see https://tailwindcss.com/docs/border-radius
             */
            &apos;rounded-ee&apos;: [{ &apos;rounded-ee&apos;: [borderRadius] }],
            /**
             * Border Radius End Start
             * @see https://tailwindcss.com/docs/border-radius
             */
            &apos;rounded-es&apos;: [{ &apos;rounded-es&apos;: [borderRadius] }],
            /**
             * Border Radius Top Left
             * @see https://tailwindcss.com/docs/border-radius
             */
            &apos;rounded-tl&apos;: [{ &apos;rounded-tl&apos;: [borderRadius] }],
            /**
             * Border Radius Top Right
             * @see https://tailwindcss.com/docs/border-radius
             */
            &apos;rounded-tr&apos;: [{ &apos;rounded-tr&apos;: [borderRadius] }],
            /**
             * Border Radius Bottom Right
             * @see https://tailwindcss.com/docs/border-radius
             */
            &apos;rounded-br&apos;: [{ &apos;rounded-br&apos;: [borderRadius] }],
            /**
             * Border Radius Bottom Left
             * @see https://tailwindcss.com/docs/border-radius
             */
            &apos;rounded-bl&apos;: [{ &apos;rounded-bl&apos;: [borderRadius] }],
            /**
             * Border Width
             * @see https://tailwindcss.com/docs/border-width
             */
            &apos;border-w&apos;: [{ border: [borderWidth] }],
            /**
             * Border Width X
             * @see https://tailwindcss.com/docs/border-width
             */
            &apos;border-w-x&apos;: [{ &apos;border-x&apos;: [borderWidth] }],
            /**
             * Border Width Y
             * @see https://tailwindcss.com/docs/border-width
             */
            &apos;border-w-y&apos;: [{ &apos;border-y&apos;: [borderWidth] }],
            /**
             * Border Width Start
             * @see https://tailwindcss.com/docs/border-width
             */
            &apos;border-w-s&apos;: [{ &apos;border-s&apos;: [borderWidth] }],
            /**
             * Border Width End
             * @see https://tailwindcss.com/docs/border-width
             */
            &apos;border-w-e&apos;: [{ &apos;border-e&apos;: [borderWidth] }],
            /**
             * Border Width Top
             * @see https://tailwindcss.com/docs/border-width
             */
            &apos;border-w-t&apos;: [{ &apos;border-t&apos;: [borderWidth] }],
            /**
             * Border Width Right
             * @see https://tailwindcss.com/docs/border-width
             */
            &apos;border-w-r&apos;: [{ &apos;border-r&apos;: [borderWidth] }],
            /**
             * Border Width Bottom
             * @see https://tailwindcss.com/docs/border-width
             */
            &apos;border-w-b&apos;: [{ &apos;border-b&apos;: [borderWidth] }],
            /**
             * Border Width Left
             * @see https://tailwindcss.com/docs/border-width
             */
            &apos;border-w-l&apos;: [{ &apos;border-l&apos;: [borderWidth] }],
            /**
             * Border Opacity
             * @see https://tailwindcss.com/docs/border-opacity
             */
            &apos;border-opacity&apos;: [{ &apos;border-opacity&apos;: [opacity] }],
            /**
             * Border Style
             * @see https://tailwindcss.com/docs/border-style
             */
            &apos;border-style&apos;: [{ border: [...getLineStyles(), &apos;hidden&apos;] }],
            /**
             * Divide Width X
             * @see https://tailwindcss.com/docs/divide-width
             */
            &apos;divide-x&apos;: [{ &apos;divide-x&apos;: [borderWidth] }],
            /**
             * Divide Width X Reverse
             * @see https://tailwindcss.com/docs/divide-width
             */
            &apos;divide-x-reverse&apos;: [&apos;divide-x-reverse&apos;],
            /**
             * Divide Width Y
             * @see https://tailwindcss.com/docs/divide-width
             */
            &apos;divide-y&apos;: [{ &apos;divide-y&apos;: [borderWidth] }],
            /**
             * Divide Width Y Reverse
             * @see https://tailwindcss.com/docs/divide-width
             */
            &apos;divide-y-reverse&apos;: [&apos;divide-y-reverse&apos;],
            /**
             * Divide Opacity
             * @see https://tailwindcss.com/docs/divide-opacity
             */
            &apos;divide-opacity&apos;: [{ &apos;divide-opacity&apos;: [opacity] }],
            /**
             * Divide Style
             * @see https://tailwindcss.com/docs/divide-style
             */
            &apos;divide-style&apos;: [{ divide: getLineStyles() }],
            /**
             * Border Color
             * @see https://tailwindcss.com/docs/border-color
             */
            &apos;border-color&apos;: [{ border: [borderColor] }],
            /**
             * Border Color X
             * @see https://tailwindcss.com/docs/border-color
             */
            &apos;border-color-x&apos;: [{ &apos;border-x&apos;: [borderColor] }],
            /**
             * Border Color Y
             * @see https://tailwindcss.com/docs/border-color
             */
            &apos;border-color-y&apos;: [{ &apos;border-y&apos;: [borderColor] }],
            /**
             * Border Color S
             * @see https://tailwindcss.com/docs/border-color
             */
            &apos;border-color-s&apos;: [{ &apos;border-s&apos;: [borderColor] }],
            /**
             * Border Color E
             * @see https://tailwindcss.com/docs/border-color
             */
            &apos;border-color-e&apos;: [{ &apos;border-e&apos;: [borderColor] }],
            /**
             * Border Color Top
             * @see https://tailwindcss.com/docs/border-color
             */
            &apos;border-color-t&apos;: [{ &apos;border-t&apos;: [borderColor] }],
            /**
             * Border Color Right
             * @see https://tailwindcss.com/docs/border-color
             */
            &apos;border-color-r&apos;: [{ &apos;border-r&apos;: [borderColor] }],
            /**
             * Border Color Bottom
             * @see https://tailwindcss.com/docs/border-color
             */
            &apos;border-color-b&apos;: [{ &apos;border-b&apos;: [borderColor] }],
            /**
             * Border Color Left
             * @see https://tailwindcss.com/docs/border-color
             */
            &apos;border-color-l&apos;: [{ &apos;border-l&apos;: [borderColor] }],
            /**
             * Divide Color
             * @see https://tailwindcss.com/docs/divide-color
             */
            &apos;divide-color&apos;: [{ divide: [borderColor] }],
            /**
             * Outline Style
             * @see https://tailwindcss.com/docs/outline-style
             */
            &apos;outline-style&apos;: [{ outline: [&apos;&apos;, ...getLineStyles()] }],
            /**
             * Outline Offset
             * @see https://tailwindcss.com/docs/outline-offset
             */
            &apos;outline-offset&apos;: [{ &apos;outline-offset&apos;: [isLength, isArbitraryValue] }],
            /**
             * Outline Width
             * @see https://tailwindcss.com/docs/outline-width
             */
            &apos;outline-w&apos;: [{ outline: [isLength, isArbitraryLength] }],
            /**
             * Outline Color
             * @see https://tailwindcss.com/docs/outline-color
             */
            &apos;outline-color&apos;: [{ outline: [colors] }],
            /**
             * Ring Width
             * @see https://tailwindcss.com/docs/ring-width
             */
            &apos;ring-w&apos;: [{ ring: getLengthWithEmptyAndArbitrary() }],
            /**
             * Ring Width Inset
             * @see https://tailwindcss.com/docs/ring-width
             */
            &apos;ring-w-inset&apos;: [&apos;ring-inset&apos;],
            /**
             * Ring Color
             * @see https://tailwindcss.com/docs/ring-color
             */
            &apos;ring-color&apos;: [{ ring: [colors] }],
            /**
             * Ring Opacity
             * @see https://tailwindcss.com/docs/ring-opacity
             */
            &apos;ring-opacity&apos;: [{ &apos;ring-opacity&apos;: [opacity] }],
            /**
             * Ring Offset Width
             * @see https://tailwindcss.com/docs/ring-offset-width
             */
            &apos;ring-offset-w&apos;: [{ &apos;ring-offset&apos;: [isLength, isArbitraryLength] }],
            /**
             * Ring Offset Color
             * @see https://tailwindcss.com/docs/ring-offset-color
             */
            &apos;ring-offset-color&apos;: [{ &apos;ring-offset&apos;: [colors] }],
            // Effects
            /**
             * Box Shadow
             * @see https://tailwindcss.com/docs/box-shadow
             */
            shadow: [{ shadow: [&apos;&apos;, &apos;inner&apos;, &apos;none&apos;, isTshirtSize, isArbitraryShadow] }],
            /**
             * Box Shadow Color
             * @see https://tailwindcss.com/docs/box-shadow-color
             */
            &apos;shadow-color&apos;: [{ shadow: [isAny] }],
            /**
             * Opacity
             * @see https://tailwindcss.com/docs/opacity
             */
            opacity: [{ opacity: [opacity] }],
            /**
             * Mix Blend Mode
             * @see https://tailwindcss.com/docs/mix-blend-mode
             */
            &apos;mix-blend&apos;: [{ &apos;mix-blend&apos;: [...getBlendModes(), &apos;plus-lighter&apos;, &apos;plus-darker&apos;] }],
            /**
             * Background Blend Mode
             * @see https://tailwindcss.com/docs/background-blend-mode
             */
            &apos;bg-blend&apos;: [{ &apos;bg-blend&apos;: getBlendModes() }],
            // Filters
            /**
             * Filter
             * @deprecated since Tailwind CSS v3.0.0
             * @see https://tailwindcss.com/docs/filter
             */
            filter: [{ filter: [&apos;&apos;, &apos;none&apos;] }],
            /**
             * Blur
             * @see https://tailwindcss.com/docs/blur
             */
            blur: [{ blur: [blur] }],
            /**
             * Brightness
             * @see https://tailwindcss.com/docs/brightness
             */
            brightness: [{ brightness: [brightness] }],
            /**
             * Contrast
             * @see https://tailwindcss.com/docs/contrast
             */
            contrast: [{ contrast: [contrast] }],
            /**
             * Drop Shadow
             * @see https://tailwindcss.com/docs/drop-shadow
             */
            &apos;drop-shadow&apos;: [{ &apos;drop-shadow&apos;: [&apos;&apos;, &apos;none&apos;, isTshirtSize, isArbitraryValue] }],
            /**
             * Grayscale
             * @see https://tailwindcss.com/docs/grayscale
             */
            grayscale: [{ grayscale: [grayscale] }],
            /**
             * Hue Rotate
             * @see https://tailwindcss.com/docs/hue-rotate
             */
            &apos;hue-rotate&apos;: [{ &apos;hue-rotate&apos;: [hueRotate] }],
            /**
             * Invert
             * @see https://tailwindcss.com/docs/invert
             */
            invert: [{ invert: [invert] }],
            /**
             * Saturate
             * @see https://tailwindcss.com/docs/saturate
             */
            saturate: [{ saturate: [saturate] }],
            /**
             * Sepia
             * @see https://tailwindcss.com/docs/sepia
             */
            sepia: [{ sepia: [sepia] }],
            /**
             * Backdrop Filter
             * @deprecated since Tailwind CSS v3.0.0
             * @see https://tailwindcss.com/docs/backdrop-filter
             */
            &apos;backdrop-filter&apos;: [{ &apos;backdrop-filter&apos;: [&apos;&apos;, &apos;none&apos;] }],
            /**
             * Backdrop Blur
             * @see https://tailwindcss.com/docs/backdrop-blur
             */
            &apos;backdrop-blur&apos;: [{ &apos;backdrop-blur&apos;: [blur] }],
            /**
             * Backdrop Brightness
             * @see https://tailwindcss.com/docs/backdrop-brightness
             */
            &apos;backdrop-brightness&apos;: [{ &apos;backdrop-brightness&apos;: [brightness] }],
            /**
             * Backdrop Contrast
             * @see https://tailwindcss.com/docs/backdrop-contrast
             */
            &apos;backdrop-contrast&apos;: [{ &apos;backdrop-contrast&apos;: [contrast] }],
            /**
             * Backdrop Grayscale
             * @see https://tailwindcss.com/docs/backdrop-grayscale
             */
            &apos;backdrop-grayscale&apos;: [{ &apos;backdrop-grayscale&apos;: [grayscale] }],
            /**
             * Backdrop Hue Rotate
             * @see https://tailwindcss.com/docs/backdrop-hue-rotate
             */
            &apos;backdrop-hue-rotate&apos;: [{ &apos;backdrop-hue-rotate&apos;: [hueRotate] }],
            /**
             * Backdrop Invert
             * @see https://tailwindcss.com/docs/backdrop-invert
             */
            &apos;backdrop-invert&apos;: [{ &apos;backdrop-invert&apos;: [invert] }],
            /**
             * Backdrop Opacity
             * @see https://tailwindcss.com/docs/backdrop-opacity
             */
            &apos;backdrop-opacity&apos;: [{ &apos;backdrop-opacity&apos;: [opacity] }],
            /**
             * Backdrop Saturate
             * @see https://tailwindcss.com/docs/backdrop-saturate
             */
            &apos;backdrop-saturate&apos;: [{ &apos;backdrop-saturate&apos;: [saturate] }],
            /**
             * Backdrop Sepia
             * @see https://tailwindcss.com/docs/backdrop-sepia
             */
            &apos;backdrop-sepia&apos;: [{ &apos;backdrop-sepia&apos;: [sepia] }],
            // Tables
            /**
             * Border Collapse
             * @see https://tailwindcss.com/docs/border-collapse
             */
            &apos;border-collapse&apos;: [{ border: [&apos;collapse&apos;, &apos;separate&apos;] }],
            /**
             * Border Spacing
             * @see https://tailwindcss.com/docs/border-spacing
             */
            &apos;border-spacing&apos;: [{ &apos;border-spacing&apos;: [borderSpacing] }],
            /**
             * Border Spacing X
             * @see https://tailwindcss.com/docs/border-spacing
             */
            &apos;border-spacing-x&apos;: [{ &apos;border-spacing-x&apos;: [borderSpacing] }],
            /**
             * Border Spacing Y
             * @see https://tailwindcss.com/docs/border-spacing
             */
            &apos;border-spacing-y&apos;: [{ &apos;border-spacing-y&apos;: [borderSpacing] }],
            /**
             * Table Layout
             * @see https://tailwindcss.com/docs/table-layout
             */
            &apos;table-layout&apos;: [{ table: [&apos;auto&apos;, &apos;fixed&apos;] }],
            /**
             * Caption Side
             * @see https://tailwindcss.com/docs/caption-side
             */
            caption: [{ caption: [&apos;top&apos;, &apos;bottom&apos;] }],
            // Transitions and Animation
            /**
             * Tranisition Property
             * @see https://tailwindcss.com/docs/transition-property
             */
            transition: [
                {
                    transition: [
                        &apos;none&apos;,
                        &apos;all&apos;,
                        &apos;&apos;,
                        &apos;colors&apos;,
                        &apos;opacity&apos;,
                        &apos;shadow&apos;,
                        &apos;transform&apos;,
                        isArbitraryValue,
                    ],
                },
            ],
            /**
             * Transition Duration
             * @see https://tailwindcss.com/docs/transition-duration
             */
            duration: [{ duration: getNumberAndArbitrary() }],
            /**
             * Transition Timing Function
             * @see https://tailwindcss.com/docs/transition-timing-function
             */
            ease: [{ ease: [&apos;linear&apos;, &apos;in&apos;, &apos;out&apos;, &apos;in-out&apos;, isArbitraryValue] }],
            /**
             * Transition Delay
             * @see https://tailwindcss.com/docs/transition-delay
             */
            delay: [{ delay: getNumberAndArbitrary() }],
            /**
             * Animation
             * @see https://tailwindcss.com/docs/animation
             */
            animate: [{ animate: [&apos;none&apos;, &apos;spin&apos;, &apos;ping&apos;, &apos;pulse&apos;, &apos;bounce&apos;, isArbitraryValue] }],
            // Transforms
            /**
             * Transform
             * @see https://tailwindcss.com/docs/transform
             */
            transform: [{ transform: [&apos;&apos;, &apos;gpu&apos;, &apos;none&apos;] }],
            /**
             * Scale
             * @see https://tailwindcss.com/docs/scale
             */
            scale: [{ scale: [scale] }],
            /**
             * Scale X
             * @see https://tailwindcss.com/docs/scale
             */
            &apos;scale-x&apos;: [{ &apos;scale-x&apos;: [scale] }],
            /**
             * Scale Y
             * @see https://tailwindcss.com/docs/scale
             */
            &apos;scale-y&apos;: [{ &apos;scale-y&apos;: [scale] }],
            /**
             * Rotate
             * @see https://tailwindcss.com/docs/rotate
             */
            rotate: [{ rotate: [isInteger, isArbitraryValue] }],
            /**
             * Translate X
             * @see https://tailwindcss.com/docs/translate
             */
            &apos;translate-x&apos;: [{ &apos;translate-x&apos;: [translate] }],
            /**
             * Translate Y
             * @see https://tailwindcss.com/docs/translate
             */
            &apos;translate-y&apos;: [{ &apos;translate-y&apos;: [translate] }],
            /**
             * Skew X
             * @see https://tailwindcss.com/docs/skew
             */
            &apos;skew-x&apos;: [{ &apos;skew-x&apos;: [skew] }],
            /**
             * Skew Y
             * @see https://tailwindcss.com/docs/skew
             */
            &apos;skew-y&apos;: [{ &apos;skew-y&apos;: [skew] }],
            /**
             * Transform Origin
             * @see https://tailwindcss.com/docs/transform-origin
             */
            &apos;transform-origin&apos;: [
                {
                    origin: [
                        &apos;center&apos;,
                        &apos;top&apos;,
                        &apos;top-right&apos;,
                        &apos;right&apos;,
                        &apos;bottom-right&apos;,
                        &apos;bottom&apos;,
                        &apos;bottom-left&apos;,
                        &apos;left&apos;,
                        &apos;top-left&apos;,
                        isArbitraryValue,
                    ],
                },
            ],
            // Interactivity
            /**
             * Accent Color
             * @see https://tailwindcss.com/docs/accent-color
             */
            accent: [{ accent: [&apos;auto&apos;, colors] }],
            /**
             * Appearance
             * @see https://tailwindcss.com/docs/appearance
             */
            appearance: [{ appearance: [&apos;none&apos;, &apos;auto&apos;] }],
            /**
             * Cursor
             * @see https://tailwindcss.com/docs/cursor
             */
            cursor: [
                {
                    cursor: [
                        &apos;auto&apos;,
                        &apos;default&apos;,
                        &apos;pointer&apos;,
                        &apos;wait&apos;,
                        &apos;text&apos;,
                        &apos;move&apos;,
                        &apos;help&apos;,
                        &apos;not-allowed&apos;,
                        &apos;none&apos;,
                        &apos;context-menu&apos;,
                        &apos;progress&apos;,
                        &apos;cell&apos;,
                        &apos;crosshair&apos;,
                        &apos;vertical-text&apos;,
                        &apos;alias&apos;,
                        &apos;copy&apos;,
                        &apos;no-drop&apos;,
                        &apos;grab&apos;,
                        &apos;grabbing&apos;,
                        &apos;all-scroll&apos;,
                        &apos;col-resize&apos;,
                        &apos;row-resize&apos;,
                        &apos;n-resize&apos;,
                        &apos;e-resize&apos;,
                        &apos;s-resize&apos;,
                        &apos;w-resize&apos;,
                        &apos;ne-resize&apos;,
                        &apos;nw-resize&apos;,
                        &apos;se-resize&apos;,
                        &apos;sw-resize&apos;,
                        &apos;ew-resize&apos;,
                        &apos;ns-resize&apos;,
                        &apos;nesw-resize&apos;,
                        &apos;nwse-resize&apos;,
                        &apos;zoom-in&apos;,
                        &apos;zoom-out&apos;,
                        isArbitraryValue,
                    ],
                },
            ],
            /**
             * Caret Color
             * @see https://tailwindcss.com/docs/just-in-time-mode#caret-color-utilities
             */
            &apos;caret-color&apos;: [{ caret: [colors] }],
            /**
             * Pointer Events
             * @see https://tailwindcss.com/docs/pointer-events
             */
            &apos;pointer-events&apos;: [{ &apos;pointer-events&apos;: [&apos;none&apos;, &apos;auto&apos;] }],
            /**
             * Resize
             * @see https://tailwindcss.com/docs/resize
             */
            resize: [{ resize: [&apos;none&apos;, &apos;y&apos;, &apos;x&apos;, &apos;&apos;] }],
            /**
             * Scroll Behavior
             * @see https://tailwindcss.com/docs/scroll-behavior
             */
            &apos;scroll-behavior&apos;: [{ scroll: [&apos;auto&apos;, &apos;smooth&apos;] }],
            /**
             * Scroll Margin
             * @see https://tailwindcss.com/docs/scroll-margin
             */
            &apos;scroll-m&apos;: [{ &apos;scroll-m&apos;: getSpacingWithArbitrary() }],
            /**
             * Scroll Margin X
             * @see https://tailwindcss.com/docs/scroll-margin
             */
            &apos;scroll-mx&apos;: [{ &apos;scroll-mx&apos;: getSpacingWithArbitrary() }],
            /**
             * Scroll Margin Y
             * @see https://tailwindcss.com/docs/scroll-margin
             */
            &apos;scroll-my&apos;: [{ &apos;scroll-my&apos;: getSpacingWithArbitrary() }],
            /**
             * Scroll Margin Start
             * @see https://tailwindcss.com/docs/scroll-margin
             */
            &apos;scroll-ms&apos;: [{ &apos;scroll-ms&apos;: getSpacingWithArbitrary() }],
            /**
             * Scroll Margin End
             * @see https://tailwindcss.com/docs/scroll-margin
             */
            &apos;scroll-me&apos;: [{ &apos;scroll-me&apos;: getSpacingWithArbitrary() }],
            /**
             * Scroll Margin Top
             * @see https://tailwindcss.com/docs/scroll-margin
             */
            &apos;scroll-mt&apos;: [{ &apos;scroll-mt&apos;: getSpacingWithArbitrary() }],
            /**
             * Scroll Margin Right
             * @see https://tailwindcss.com/docs/scroll-margin
             */
            &apos;scroll-mr&apos;: [{ &apos;scroll-mr&apos;: getSpacingWithArbitrary() }],
            /**
             * Scroll Margin Bottom
             * @see https://tailwindcss.com/docs/scroll-margin
             */
            &apos;scroll-mb&apos;: [{ &apos;scroll-mb&apos;: getSpacingWithArbitrary() }],
            /**
             * Scroll Margin Left
             * @see https://tailwindcss.com/docs/scroll-margin
             */
            &apos;scroll-ml&apos;: [{ &apos;scroll-ml&apos;: getSpacingWithArbitrary() }],
            /**
             * Scroll Padding
             * @see https://tailwindcss.com/docs/scroll-padding
             */
            &apos;scroll-p&apos;: [{ &apos;scroll-p&apos;: getSpacingWithArbitrary() }],
            /**
             * Scroll Padding X
             * @see https://tailwindcss.com/docs/scroll-padding
             */
            &apos;scroll-px&apos;: [{ &apos;scroll-px&apos;: getSpacingWithArbitrary() }],
            /**
             * Scroll Padding Y
             * @see https://tailwindcss.com/docs/scroll-padding
             */
            &apos;scroll-py&apos;: [{ &apos;scroll-py&apos;: getSpacingWithArbitrary() }],
            /**
             * Scroll Padding Start
             * @see https://tailwindcss.com/docs/scroll-padding
             */
            &apos;scroll-ps&apos;: [{ &apos;scroll-ps&apos;: getSpacingWithArbitrary() }],
            /**
             * Scroll Padding End
             * @see https://tailwindcss.com/docs/scroll-padding
             */
            &apos;scroll-pe&apos;: [{ &apos;scroll-pe&apos;: getSpacingWithArbitrary() }],
            /**
             * Scroll Padding Top
             * @see https://tailwindcss.com/docs/scroll-padding
             */
            &apos;scroll-pt&apos;: [{ &apos;scroll-pt&apos;: getSpacingWithArbitrary() }],
            /**
             * Scroll Padding Right
             * @see https://tailwindcss.com/docs/scroll-padding
             */
            &apos;scroll-pr&apos;: [{ &apos;scroll-pr&apos;: getSpacingWithArbitrary() }],
            /**
             * Scroll Padding Bottom
             * @see https://tailwindcss.com/docs/scroll-padding
             */
            &apos;scroll-pb&apos;: [{ &apos;scroll-pb&apos;: getSpacingWithArbitrary() }],
            /**
             * Scroll Padding Left
             * @see https://tailwindcss.com/docs/scroll-padding
             */
            &apos;scroll-pl&apos;: [{ &apos;scroll-pl&apos;: getSpacingWithArbitrary() }],
            /**
             * Scroll Snap Align
             * @see https://tailwindcss.com/docs/scroll-snap-align
             */
            &apos;snap-align&apos;: [{ snap: [&apos;start&apos;, &apos;end&apos;, &apos;center&apos;, &apos;align-none&apos;] }],
            /**
             * Scroll Snap Stop
             * @see https://tailwindcss.com/docs/scroll-snap-stop
             */
            &apos;snap-stop&apos;: [{ snap: [&apos;normal&apos;, &apos;always&apos;] }],
            /**
             * Scroll Snap Type
             * @see https://tailwindcss.com/docs/scroll-snap-type
             */
            &apos;snap-type&apos;: [{ snap: [&apos;none&apos;, &apos;x&apos;, &apos;y&apos;, &apos;both&apos;] }],
            /**
             * Scroll Snap Type Strictness
             * @see https://tailwindcss.com/docs/scroll-snap-type
             */
            &apos;snap-strictness&apos;: [{ snap: [&apos;mandatory&apos;, &apos;proximity&apos;] }],
            /**
             * Touch Action
             * @see https://tailwindcss.com/docs/touch-action
             */
            touch: [
                {
                    touch: [&apos;auto&apos;, &apos;none&apos;, &apos;manipulation&apos;],
                },
            ],
            /**
             * Touch Action X
             * @see https://tailwindcss.com/docs/touch-action
             */
            &apos;touch-x&apos;: [
                {
                    &apos;touch-pan&apos;: [&apos;x&apos;, &apos;left&apos;, &apos;right&apos;],
                },
            ],
            /**
             * Touch Action Y
             * @see https://tailwindcss.com/docs/touch-action
             */
            &apos;touch-y&apos;: [
                {
                    &apos;touch-pan&apos;: [&apos;y&apos;, &apos;up&apos;, &apos;down&apos;],
                },
            ],
            /**
             * Touch Action Pinch Zoom
             * @see https://tailwindcss.com/docs/touch-action
             */
            &apos;touch-pz&apos;: [&apos;touch-pinch-zoom&apos;],
            /**
             * User Select
             * @see https://tailwindcss.com/docs/user-select
             */
            select: [{ select: [&apos;none&apos;, &apos;text&apos;, &apos;all&apos;, &apos;auto&apos;] }],
            /**
             * Will Change
             * @see https://tailwindcss.com/docs/will-change
             */
            &apos;will-change&apos;: [
                { &apos;will-change&apos;: [&apos;auto&apos;, &apos;scroll&apos;, &apos;contents&apos;, &apos;transform&apos;, isArbitraryValue] },
            ],
            // SVG
            /**
             * Fill
             * @see https://tailwindcss.com/docs/fill
             */
            fill: [{ fill: [colors, &apos;none&apos;] }],
            /**
             * Stroke Width
             * @see https://tailwindcss.com/docs/stroke-width
             */
            &apos;stroke-w&apos;: [{ stroke: [isLength, isArbitraryLength, isArbitraryNumber] }],
            /**
             * Stroke
             * @see https://tailwindcss.com/docs/stroke
             */
            stroke: [{ stroke: [colors, &apos;none&apos;] }],
            // Accessibility
            /**
             * Screen Readers
             * @see https://tailwindcss.com/docs/screen-readers
             */
            sr: [&apos;sr-only&apos;, &apos;not-sr-only&apos;],
            /**
             * Forced Color Adjust
             * @see https://tailwindcss.com/docs/forced-color-adjust
             */
            &apos;forced-color-adjust&apos;: [{ &apos;forced-color-adjust&apos;: [&apos;auto&apos;, &apos;none&apos;] }],
        },
        conflictingClassGroups: {
            overflow: [&apos;overflow-x&apos;, &apos;overflow-y&apos;],
            overscroll: [&apos;overscroll-x&apos;, &apos;overscroll-y&apos;],
            inset: [&apos;inset-x&apos;, &apos;inset-y&apos;, &apos;start&apos;, &apos;end&apos;, &apos;top&apos;, &apos;right&apos;, &apos;bottom&apos;, &apos;left&apos;],
            &apos;inset-x&apos;: [&apos;right&apos;, &apos;left&apos;],
            &apos;inset-y&apos;: [&apos;top&apos;, &apos;bottom&apos;],
            flex: [&apos;basis&apos;, &apos;grow&apos;, &apos;shrink&apos;],
            gap: [&apos;gap-x&apos;, &apos;gap-y&apos;],
            p: [&apos;px&apos;, &apos;py&apos;, &apos;ps&apos;, &apos;pe&apos;, &apos;pt&apos;, &apos;pr&apos;, &apos;pb&apos;, &apos;pl&apos;],
            px: [&apos;pr&apos;, &apos;pl&apos;],
            py: [&apos;pt&apos;, &apos;pb&apos;],
            m: [&apos;mx&apos;, &apos;my&apos;, &apos;ms&apos;, &apos;me&apos;, &apos;mt&apos;, &apos;mr&apos;, &apos;mb&apos;, &apos;ml&apos;],
            mx: [&apos;mr&apos;, &apos;ml&apos;],
            my: [&apos;mt&apos;, &apos;mb&apos;],
            size: [&apos;w&apos;, &apos;h&apos;],
            &apos;font-size&apos;: [&apos;leading&apos;],
            &apos;fvn-normal&apos;: [
                &apos;fvn-ordinal&apos;,
                &apos;fvn-slashed-zero&apos;,
                &apos;fvn-figure&apos;,
                &apos;fvn-spacing&apos;,
                &apos;fvn-fraction&apos;,
            ],
            &apos;fvn-ordinal&apos;: [&apos;fvn-normal&apos;],
            &apos;fvn-slashed-zero&apos;: [&apos;fvn-normal&apos;],
            &apos;fvn-figure&apos;: [&apos;fvn-normal&apos;],
            &apos;fvn-spacing&apos;: [&apos;fvn-normal&apos;],
            &apos;fvn-fraction&apos;: [&apos;fvn-normal&apos;],
            &apos;line-clamp&apos;: [&apos;display&apos;, &apos;overflow&apos;],
            rounded: [
                &apos;rounded-s&apos;,
                &apos;rounded-e&apos;,
                &apos;rounded-t&apos;,
                &apos;rounded-r&apos;,
                &apos;rounded-b&apos;,
                &apos;rounded-l&apos;,
                &apos;rounded-ss&apos;,
                &apos;rounded-se&apos;,
                &apos;rounded-ee&apos;,
                &apos;rounded-es&apos;,
                &apos;rounded-tl&apos;,
                &apos;rounded-tr&apos;,
                &apos;rounded-br&apos;,
                &apos;rounded-bl&apos;,
            ],
            &apos;rounded-s&apos;: [&apos;rounded-ss&apos;, &apos;rounded-es&apos;],
            &apos;rounded-e&apos;: [&apos;rounded-se&apos;, &apos;rounded-ee&apos;],
            &apos;rounded-t&apos;: [&apos;rounded-tl&apos;, &apos;rounded-tr&apos;],
            &apos;rounded-r&apos;: [&apos;rounded-tr&apos;, &apos;rounded-br&apos;],
            &apos;rounded-b&apos;: [&apos;rounded-br&apos;, &apos;rounded-bl&apos;],
            &apos;rounded-l&apos;: [&apos;rounded-tl&apos;, &apos;rounded-bl&apos;],
            &apos;border-spacing&apos;: [&apos;border-spacing-x&apos;, &apos;border-spacing-y&apos;],
            &apos;border-w&apos;: [
                &apos;border-w-s&apos;,
                &apos;border-w-e&apos;,
                &apos;border-w-t&apos;,
                &apos;border-w-r&apos;,
                &apos;border-w-b&apos;,
                &apos;border-w-l&apos;,
            ],
            &apos;border-w-x&apos;: [&apos;border-w-r&apos;, &apos;border-w-l&apos;],
            &apos;border-w-y&apos;: [&apos;border-w-t&apos;, &apos;border-w-b&apos;],
            &apos;border-color&apos;: [
                &apos;border-color-s&apos;,
                &apos;border-color-e&apos;,
                &apos;border-color-t&apos;,
                &apos;border-color-r&apos;,
                &apos;border-color-b&apos;,
                &apos;border-color-l&apos;,
            ],
            &apos;border-color-x&apos;: [&apos;border-color-r&apos;, &apos;border-color-l&apos;],
            &apos;border-color-y&apos;: [&apos;border-color-t&apos;, &apos;border-color-b&apos;],
            &apos;scroll-m&apos;: [
                &apos;scroll-mx&apos;,
                &apos;scroll-my&apos;,
                &apos;scroll-ms&apos;,
                &apos;scroll-me&apos;,
                &apos;scroll-mt&apos;,
                &apos;scroll-mr&apos;,
                &apos;scroll-mb&apos;,
                &apos;scroll-ml&apos;,
            ],
            &apos;scroll-mx&apos;: [&apos;scroll-mr&apos;, &apos;scroll-ml&apos;],
            &apos;scroll-my&apos;: [&apos;scroll-mt&apos;, &apos;scroll-mb&apos;],
            &apos;scroll-p&apos;: [
                &apos;scroll-px&apos;,
                &apos;scroll-py&apos;,
                &apos;scroll-ps&apos;,
                &apos;scroll-pe&apos;,
                &apos;scroll-pt&apos;,
                &apos;scroll-pr&apos;,
                &apos;scroll-pb&apos;,
                &apos;scroll-pl&apos;,
            ],
            &apos;scroll-px&apos;: [&apos;scroll-pr&apos;, &apos;scroll-pl&apos;],
            &apos;scroll-py&apos;: [&apos;scroll-pt&apos;, &apos;scroll-pb&apos;],
            touch: [&apos;touch-x&apos;, &apos;touch-y&apos;, &apos;touch-pz&apos;],
            &apos;touch-x&apos;: [&apos;touch&apos;],
            &apos;touch-y&apos;: [&apos;touch&apos;],
            &apos;touch-pz&apos;: [&apos;touch&apos;],
        },
        conflictingClassGroupModifiers: {
            &apos;font-size&apos;: [&apos;leading&apos;],
        },
    } as const satisfies Config<DefaultClassGroupIds, DefaultThemeGroupIds>
}
