// tailwindcss@3.4.16/scripts/generate-types.js
import prettier from 'prettier'
import { corePlugins } from '../src/corePlugins'
import colors from '../src/public/colors'
import defaultTheme from '../src/public/default-theme'
import fs from 'fs'
import path from 'path'
import * as types from './type-utils'

fs.writeFileSync(
  path.join(process.cwd(), 'types', 'generated', 'corePluginList.d.ts'),
  `export type CorePluginList = ${Object.keys(corePlugins)
    .map((p) => `&apos;${p}&apos;`)
    .join(&apos; | &apos;)}`
)

let colorsWithoutDeprecatedColors = Object.fromEntries(
  Object.entries(Object.getOwnPropertyDescriptors(colors))
    .filter(([_, { value }]) => {
      return typeof value !== &apos;undefined&apos;
    })
    .map(([name, definition]) => [name, definition.value])
)

let deprecatedColors = Object.entries(Object.getOwnPropertyDescriptors(colors))
  .filter(([_, { value }]) => {
    return typeof value === &apos;undefined&apos;
  })
  .map(([name, definition]) => {
    let warn = console.warn
    let messages = []
    console.warn = (...args) => messages.push(args.pop())
    definition.get()
    console.warn = warn
    let message = messages.join(&apos; &apos;).trim()
    let newColor = message.match(/renamed to `(.*)`/)[1]
    return `/** @deprecated ${message} */${name}: DefaultColors[&apos;${newColor}&apos;],`
  })
  .join(&apos;\n&apos;)

fs.writeFileSync(
  path.join(process.cwd(), &apos;types&apos;, &apos;generated&apos;, &apos;colors.d.ts&apos;),
  prettier.format(
    `export interface DefaultColors { ${JSON.stringify(colorsWithoutDeprecatedColors).slice(
      1,
      -1
    )}\n${deprecatedColors}\n}`,
    {
      semi: false,
      singleQuote: true,
      printWidth: 100,
      parser: &apos;typescript&apos;,
    }
  )
)

const defaultThemeTypes = Object.entries(defaultTheme)
  .map(([name, value]) => {
    // Special cases for slightly more accurate types
    if (name === &apos;keyframes&apos;) {
      return [name, `Record<${types.forKeys(value)}, Record<string, CSSDeclarationList>>`]
    }

    if (name === &apos;fontSize&apos;) {
      return [name, `Record<${types.forKeys(value)}, [string, { lineHeight: string }]>`]
    }

    // General cases
    if (typeof value === &apos;string&apos;) {
      return [name, `string`]
    }

    if (typeof value === &apos;function&apos;) {
      return [name, null]
    }

    if (typeof value === &apos;object&apos;) {
      if (Object.keys(value).length === 0) {
        return [name, null]
      }

      return [name, types.forValue(value)]
    }

    return [name, `unknown`]
  })
  .filter(([, type]) => type !== null)
  .map(([name, type]) => `${name}: ${type}`)
  .join(&apos;\n&apos;)

fs.writeFileSync(
  path.join(process.cwd(), &apos;types&apos;, &apos;generated&apos;, &apos;default-theme.d.ts&apos;),
  prettier.format(
    `
    type CSSDeclarationList = Record<string, string>
    export type DefaultTheme = { ${defaultThemeTypes} }
  `,
    {
      semi: false,
      singleQuote: true,
      printWidth: 100,
      parser: 'typescript',
    }
  )
)
