// eslint-plugin-react@15.3.3/index.js
'use strict';

const fromEntries = require('object.fromentries');
const entries = require('object.entries');

const allRules = require('./lib/rules');

function filterRules(rules, predicate) {
  return fromEntries(entries(rules).filter((entry) => predicate(entry[1])));
}

/**
 * @param {object} rules - rules object mapping rule name to rule module
 * @returns {Record<string, SEVERITY_ERROR | 'error'>}
 */
function configureAsError(rules) {
  return fromEntries(Object.keys(rules).map((key) => [`react/${key}`, 2]));
}

/** @type {Partial<typeof allRules>} */
const activeRules = filterRules(allRules, (rule) => !rule.meta.deprecated);
/** @type {Record<keyof typeof activeRules, 2 | 'error'>} */
const activeRulesConfig = configureAsError(activeRules);

/** @type {Partial<typeof allRules>} */
const deprecatedRules = filterRules(allRules, (rule) => rule.meta.deprecated);

/** @type {[&apos;react&apos;]} */
// for legacy config system
const plugins = [
  &apos;react&apos;,
];

// TODO: with TS 4.5+, inline this
const SEVERITY_ERROR = /** @type {2} */ (2);
const SEVERITY_OFF = /** @type {0} */ (0);

const configs = {
  recommended: {
    plugins,
    parserOptions: {
      ecmaFeatures: {
        jsx: true,
      },
    },
    rules: {
      &apos;react/display-name&apos;: SEVERITY_ERROR,
      &apos;react/jsx-key&apos;: SEVERITY_ERROR,
      &apos;react/jsx-no-comment-textnodes&apos;: SEVERITY_ERROR,
      &apos;react/jsx-no-duplicate-props&apos;: SEVERITY_ERROR,
      &apos;react/jsx-no-target-blank&apos;: SEVERITY_ERROR,
      &apos;react/jsx-no-undef&apos;: SEVERITY_ERROR,
      &apos;react/jsx-uses-react&apos;: SEVERITY_ERROR,
      &apos;react/jsx-uses-vars&apos;: SEVERITY_ERROR,
      &apos;react/no-children-prop&apos;: SEVERITY_ERROR,
      &apos;react/no-danger-with-children&apos;: SEVERITY_ERROR,
      &apos;react/no-deprecated&apos;: SEVERITY_ERROR,
      &apos;react/no-direct-mutation-state&apos;: SEVERITY_ERROR,
      &apos;react/no-find-dom-node&apos;: SEVERITY_ERROR,
      &apos;react/no-is-mounted&apos;: SEVERITY_ERROR,
      &apos;react/no-render-return-value&apos;: SEVERITY_ERROR,
      &apos;react/no-string-refs&apos;: SEVERITY_ERROR,
      &apos;react/no-unescaped-entities&apos;: SEVERITY_ERROR,
      &apos;react/no-unknown-property&apos;: SEVERITY_ERROR,
      &apos;react/no-unsafe&apos;: SEVERITY_OFF,
      &apos;react/prop-types&apos;: SEVERITY_ERROR,
      &apos;react/react-in-jsx-scope&apos;: SEVERITY_ERROR,
      &apos;react/require-render-return&apos;: SEVERITY_ERROR,
    },
  },
  all: {
    plugins,
    parserOptions: {
      ecmaFeatures: {
        jsx: true,
      },
    },
    rules: activeRulesConfig,
  },
  &apos;jsx-runtime&apos;: {
    plugins,
    parserOptions: {
      ecmaFeatures: {
        jsx: true,
      },
      jsxPragma: null, // for @typescript/eslint-parser
    },
    rules: {
      &apos;react/react-in-jsx-scope&apos;: SEVERITY_OFF,
      &apos;react/jsx-uses-react&apos;: SEVERITY_OFF,
    },
  },
  flat: /** @type {Record<string, ReactFlatConfig>} */ ({
    __proto__: null,
  }),
};

/** @typedef {{ plugins: { react: typeof plugin }, rules: import(&apos;eslint&apos;).Linter.RulesRecord, languageOptions: { parserOptions: import(&apos;eslint&apos;).Linter.ParserOptions } }} ReactFlatConfig */

/** @type {{ deprecatedRules: typeof deprecatedRules, rules: typeof allRules, configs: typeof configs & { flat: Record<string, ReactFlatConfig> }}} */
const plugin = {
  deprecatedRules,
  rules: allRules,
  configs,
};

Object.assign(configs.flat, {
  recommended: {
    plugins: { react: plugin },
    rules: configs.recommended.rules,
    languageOptions: { parserOptions: configs.recommended.parserOptions },
  },
  all: {
    plugins: { react: plugin },
    rules: configs.all.rules,
    languageOptions: { parserOptions: configs.all.parserOptions },
  },
  'jsx-runtime': {
    plugins: { react: plugin },
    rules: configs['jsx-runtime'].rules,
    languageOptions: { parserOptions: configs['jsx-runtime'].parserOptions },
  },
});

module.exports = plugin;
