// @hookform/resolvers@3.10.0/nope/src/__tests__/__fixtures__/data.ts
import Nope from 'nope-validator';
import { Field, InternalFieldName } from 'react-hook-form';

export const schema = Nope.object().shape({
  username: Nope.string().regex(/^\w+$/).min(2).max(30).required(),
  password: Nope.string()
    .regex(new RegExp('.*[A-Z].*'), 'One uppercase character')
    .regex(new RegExp('.*[a-z].*'), 'One lowercase character')
    .regex(new RegExp('.*\\d.*'), 'One number')
    .regex(
      new RegExp('.*[`~<>?,./!@#$%^&*()\\-_+=&quot;\&apos;|{}\\[\\];:\\\\].*&apos;),
      &apos;One special character&apos;,
    )
    .min(8, &apos;Must be at least 8 characters in length&apos;)
    .required(&apos;New Password is required&apos;),
  repeatPassword: Nope.string()
    .oneOf([Nope.ref(&apos;password&apos;)], &quot;Passwords don&apos;t match&quot;)
    .required(),
  accessToken: Nope.string(),
  birthYear: Nope.number().min(1900).max(2013),
  email: Nope.string().email(),
  tags: Nope.array().of(Nope.string()).required(),
  enabled: Nope.boolean(),
  like: Nope.object().shape({
    id: Nope.number().required(),
    name: Nope.string().atLeast(4).required(),
  }),
});

export const validData = {
  username: &apos;Doe&apos;,
  password: &apos;Password123_&apos;,
  repeatPassword: &apos;Password123_&apos;,
  birthYear: 2000,
  email: &apos;john@doe.com&apos;,
  tags: [&apos;tag1&apos;, &apos;tag2&apos;],
  enabled: true,
  accessToken: &apos;accessToken&apos;,
  like: {
    id: 1,
    name: &apos;name&apos;,
  },
};

export const invalidData = {
  password: &apos;___&apos;,
  email: &apos;&apos;,
  birthYear: &apos;birthYear&apos;,
  like: { id: &apos;z&apos; },
  tags: [1, 2, 3],
};

export const fields: Record<InternalFieldName, Field['_f']> = {
  username: {
    ref: { name: 'username' },
    name: 'username',
  },
  password: {
    ref: { name: 'password' },
    name: 'password',
  },
  email: {
    ref: { name: 'email' },
    name: 'email',
  },
  birthday: {
    ref: { name: 'birthday' },
    name: 'birthday',
  },
};
