// @hookform/resolvers@3.10.0/typanion/src/__tests__/__fixtures__/data.ts
import { Field, InternalFieldName } from 'react-hook-form';
import * as t from 'typanion';

export const isSchema = t.isObject({
  username: t.applyCascade(t.isString(), [
    t.matchesRegExp(/^\w+$/),
    t.hasMinLength(2),
    t.hasMaxLength(30),
  ]),
  password: t.applyCascade(t.isString(), [
    t.matchesRegExp(new RegExp('.*[A-Z].*')), // one uppercase character
    t.matchesRegExp(new RegExp('.*[a-z].*')), // one lowercase character
    t.matchesRegExp(new RegExp('.*\\d.*')), // one number
    t.matchesRegExp(
      new RegExp('.*[`~<>?,./!@#$%^&*()\\-_+=&quot;\&apos;|{}\\[\\];:\\\\].*&apos;),
    ), // one special character
    t.hasMinLength(8), // Must be at least 8 characters in length
  ]),
  repeatPassword: t.applyCascade(t.isString(), [
    t.matchesRegExp(new RegExp(&apos;.*[A-Z].*&apos;)), // one uppercase character
    t.matchesRegExp(new RegExp(&apos;.*[a-z].*&apos;)), // one lowercase character
    t.matchesRegExp(new RegExp(&apos;.*\\d.*&apos;)), // one number
    t.matchesRegExp(
      new RegExp(&apos;.*[`~<>?,./!@#$%^&*()\\-_+=&quot;\&apos;|{}\\[\\];:\\\\].*&apos;),
    ), // one special character
    t.hasMinLength(8), // Must be at least 8 characters in length
  ]),
  accessToken: t.isString(),
  birthYear: t.applyCascade(t.isNumber(), [
    t.isInteger(),
    t.isInInclusiveRange(1900, 2013),
  ]),
  email: t.applyCascade(t.isString(), [t.matchesRegExp(/^\S+@\S+$/)]),
  tags: t.isArray(t.isString()),
  enabled: t.isBoolean(),
  like: t.isObject({
    id: t.applyCascade(t.isNumber(), [t.isInteger(), t.isPositive()]),
    name: t.applyCascade(t.isString(), [t.hasMinLength(4)]),
  }),
});

export const validData = {
  username: &apos;Doe&apos;,
  password: &apos;Password123_&apos;,
  repeatPassword: &apos;Password123_&apos;,
  birthYear: 2000,
  email: &apos;john@doe.com&apos;,
  tags: [&apos;tag1&apos;, &apos;tag2&apos;],
  enabled: true,
  accessToken: &apos;accessToken&apos;,
  like: {
    id: 1,
    name: &apos;name&apos;,
  },
};

export const invalidData = {
  password: &apos;___&apos;,
  email: &apos;&apos;,
  birthYear: &apos;birthYear&apos;,
  like: { id: &apos;z&apos; },
  tags: [1, 2, 3],
};

export const fields: Record<InternalFieldName, Field['_f']> = {
  username: {
    ref: { name: 'username' },
    name: 'username',
  },
  password: {
    ref: { name: 'password' },
    name: 'password',
  },
  email: {
    ref: { name: 'email' },
    name: 'email',
  },
  birthday: {
    ref: { name: 'birthday' },
    name: 'birthday',
  },
};
