// @hookform/resolvers@3.10.0/class-validator/src/__tests__/__fixtures__/data.ts
import 'reflect-metadata';
import { Type } from 'class-transformer';
import {
  IsEmail,
  IsNotEmpty,
  Length,
  Matches,
  Max,
  Min,
  ValidateNested,
} from 'class-validator';
import { Field, InternalFieldName } from 'react-hook-form';

class Like {
  @IsNotEmpty()
  id: number;

  @Length(4)
  name: string;
}

export class Schema {
  @Matches(/^\w+$/)
  @Length(3, 30)
  username: string;

  @Matches(/^[a-zA-Z0-9]{3,30}/)
  password: string;

  @Min(1900)
  @Max(2013)
  birthYear: number;

  @IsEmail()
  email: string;

  accessToken: string;

  tags: string[];

  enabled: boolean;

  @ValidateNested({ each: true })
  @Type(() => Like)
  like: Like[];
}

export const validData: Schema = {
  username: &apos;Doe&apos;,
  password: &apos;Password123&apos;,
  birthYear: 2000,
  email: &apos;john@doe.com&apos;,
  tags: [&apos;tag1&apos;, &apos;tag2&apos;],
  enabled: true,
  accessToken: &apos;accessToken&apos;,
  like: [
    {
      id: 1,
      name: &apos;name&apos;,
    },
  ],
};

export const invalidData = {
  password: &apos;___&apos;,
  email: &apos;&apos;,
  birthYear: &apos;birthYear&apos;,
  like: [{ id: &apos;z&apos; }],
};

export const fields: Record<InternalFieldName, Field['_f']> = {
  username: {
    ref: { name: 'username' },
    name: 'username',
  },
  password: {
    ref: { name: 'password' },
    name: 'password',
  },
  email: {
    ref: { name: 'email' },
    name: 'email',
  },
  birthday: {
    ref: { name: 'birthday' },
    name: 'birthday',
  },
};
