// date-fns@4.1.0/parse/_lib/parsers/LocalDayParser.js
import { setDay } from "../../../setDay.js";
import { Parser } from "../Parser.js";

import { mapValue, parseNDigits } from "../utils.js";

// Local day of week
export class LocalDayParser extends Parser {
  priority = 90;
  parse(dateString, token, match, options) {
    const valueCallback = (value) => {
      // We want here floor instead of trunc, so we get -7 for value 0 instead of 0
      const wholeWeekDays = Math.floor((value - 1) / 7) * 7;
      return ((value + options.weekStartsOn + 6) % 7) + wholeWeekDays;
    };

    switch (token) {
      // 3
      case &quot;e&quot;:
      case &quot;ee&quot;: // 03
        return mapValue(parseNDigits(token.length, dateString), valueCallback);
      // 3rd
      case &quot;eo&quot;:
        return mapValue(
          match.ordinalNumber(dateString, {
            unit: &quot;day&quot;,
          }),
          valueCallback,
        );
      // Tue
      case &quot;eee&quot;:
        return (
          match.day(dateString, {
            width: &quot;abbreviated&quot;,
            context: &quot;formatting&quot;,
          }) ||
          match.day(dateString, { width: &quot;short&quot;, context: &quot;formatting&quot; }) ||
          match.day(dateString, { width: &quot;narrow&quot;, context: &quot;formatting&quot; })
        );

      // T
      case &quot;eeeee&quot;:
        return match.day(dateString, {
          width: &quot;narrow&quot;,
          context: &quot;formatting&quot;,
        });
      // Tu
      case &quot;eeeeee&quot;:
        return (
          match.day(dateString, { width: &quot;short&quot;, context: &quot;formatting&quot; }) ||
          match.day(dateString, { width: &quot;narrow&quot;, context: &quot;formatting&quot; })
        );

      // Tuesday
      case &quot;eeee&quot;:
      default:
        return (
          match.day(dateString, { width: &quot;wide&quot;, context: &quot;formatting&quot; }) ||
          match.day(dateString, {
            width: &quot;abbreviated&quot;,
            context: &quot;formatting&quot;,
          }) ||
          match.day(dateString, { width: &quot;short&quot;, context: &quot;formatting&quot; }) ||
          match.day(dateString, { width: &quot;narrow&quot;, context: &quot;formatting&quot; })
        );
    }
  }

  validate(_date, value) {
    return value >= 0 && value <= 6;
  }

  set(date, _flags, value, options) {
    date = setDay(date, value, options);
    date.setHours(0, 0, 0, 0);
    return date;
  }

  incompatibleTokens = [
    "y",
    "R",
    "u",
    "q",
    "Q",
    "M",
    "L",
    "I",
    "d",
    "D",
    "E",
    "i",
    "c",
    "t",
    "T",
  ];
}
