// date-fns@4.1.0/parse/_lib/parsers/StandAloneMonthParser.js
import { numericPatterns } from "../constants.js";
import { Parser } from "../Parser.js";

import { mapValue, parseNDigits, parseNumericPattern } from "../utils.js";

export class StandAloneMonthParser extends Parser {
  priority = 110;

  parse(dateString, token, match) {
    const valueCallback = (value) => value - 1;

    switch (token) {
      // 1, 2, ..., 12
      case &quot;L&quot;:
        return mapValue(
          parseNumericPattern(numericPatterns.month, dateString),
          valueCallback,
        );
      // 01, 02, ..., 12
      case &quot;LL&quot;:
        return mapValue(parseNDigits(2, dateString), valueCallback);
      // 1st, 2nd, ..., 12th
      case &quot;Lo&quot;:
        return mapValue(
          match.ordinalNumber(dateString, {
            unit: &quot;month&quot;,
          }),
          valueCallback,
        );
      // Jan, Feb, ..., Dec
      case &quot;LLL&quot;:
        return (
          match.month(dateString, {
            width: &quot;abbreviated&quot;,
            context: &quot;standalone&quot;,
          }) ||
          match.month(dateString, { width: &quot;narrow&quot;, context: &quot;standalone&quot; })
        );

      // J, F, ..., D
      case &quot;LLLLL&quot;:
        return match.month(dateString, {
          width: &quot;narrow&quot;,
          context: &quot;standalone&quot;,
        });
      // January, February, ..., December
      case &quot;LLLL&quot;:
      default:
        return (
          match.month(dateString, { width: &quot;wide&quot;, context: &quot;standalone&quot; }) ||
          match.month(dateString, {
            width: &quot;abbreviated&quot;,
            context: &quot;standalone&quot;,
          }) ||
          match.month(dateString, { width: &quot;narrow&quot;, context: &quot;standalone&quot; })
        );
    }
  }

  validate(_date, value) {
    return value >= 0 && value <= 11;
  }

  set(date, _flags, value) {
    date.setMonth(value, 1);
    date.setHours(0, 0, 0, 0);
    return date;
  }

  incompatibleTokens = [
    "Y",
    "R",
    "q",
    "Q",
    "M",
    "w",
    "I",
    "D",
    "i",
    "e",
    "c",
    "t",
    "T",
  ];
}
