// @hookform/resolvers@3.10.0/typebox/src/__tests__/__fixtures__/data.ts
import { Static, Type } from '@sinclair/typebox';
import { Field, InternalFieldName } from 'react-hook-form';

export const schema = Type.Object({
  username: Type.String({
    minLength: 3,
    maxLength: 30,
    pattern: '^[a-zA-Z0-9]+$',
  }),
  password: Type.String({
    minLength: 8,
    pattern: '^(.*[A-Za-z\\d].*)$',
  }),
  repeatPassword: Type.String(),
  accessToken: Type.Union([Type.String(), Type.Number()]),
  birthYear: Type.Optional(
    Type.Number({
      minimum: 1900,
      maximum: 2013,
    }),
  ),
  email: Type.Optional(
    Type.RegExp(
      /^[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*@(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?$/i,
    ),
  ),
  tags: Type.Array(Type.String()),
  enabled: Type.Boolean(),
  like: Type.Optional(
    Type.Array(
      Type.Object({
        id: Type.Number(),
        name: Type.String({ minLength: 4, maxLength: 4 }),
      }),
    ),
  ),
  dateStr: Type.Date(),
});

export const validData: Static<typeof schema> = {
  username: &apos;Doe&apos;,
  password: &apos;Password123_&apos;,
  repeatPassword: &apos;Password123_&apos;,
  birthYear: 2000,
  email: &apos;google@gmail.com&apos;,
  tags: [&apos;tag1&apos;, &apos;tag2&apos;],
  enabled: true,
  accessToken: &apos;accessToken&apos;,
  like: [
    {
      id: 1,
      name: &apos;name&apos;,
    },
  ],
  dateStr: new Date(),
};

export const invalidData = {
  password: &apos;___&apos;,
  email: &apos;&apos;,
  birthYear: &apos;birthYear&apos;,
  like: [{ id: &apos;z&apos; }],
  url: &apos;abc&apos;,
};

export const fields: Record<InternalFieldName, Field['_f']> = {
  username: {
    ref: { name: 'username' },
    name: 'username',
  },
  password: {
    ref: { name: 'password' },
    name: 'password',
  },
  email: {
    ref: { name: 'email' },
    name: 'email',
  },
  birthday: {
    ref: { name: 'birthday' },
    name: 'birthday',
  },
};
