// eslint-plugin-react@7.37.5/lib/rules/forbid-elements.js
/**
 * @fileoverview Forbid certain elements
 * @author Kenneth Chung
 */

'use strict';

const has = require('hasown');
const docsUrl = require('../util/docsUrl');
const getText = require('../util/eslint').getText;
const isCreateElement = require('../util/isCreateElement');
const report = require('../util/report');

// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------

const messages = {
  forbiddenElement: '<{{element}}> is forbidden&apos;,
  forbiddenElement_message: &apos;<{{element}}> is forbidden, {{message}}&apos;,
};

/** @type {import(&apos;eslint&apos;).Rule.RuleModule} */
module.exports = {
  meta: {
    docs: {
      description: &apos;Disallow certain elements&apos;,
      category: &apos;Best Practices&apos;,
      recommended: false,
      url: docsUrl(&apos;forbid-elements&apos;),
    },

    messages,

    schema: [{
      type: &apos;object&apos;,
      properties: {
        forbid: {
          type: &apos;array&apos;,
          items: {
            anyOf: [
              { type: &apos;string&apos; },
              {
                type: &apos;object&apos;,
                properties: {
                  element: { type: &apos;string&apos; },
                  message: { type: &apos;string&apos; },
                },
                required: [&apos;element&apos;],
                additionalProperties: false,
              },
            ],
          },
        },
      },
      additionalProperties: false,
    }],
  },

  create(context) {
    const configuration = context.options[0] || {};
    const forbidConfiguration = configuration.forbid || [];

    /** @type {Record<string, { element: string, message?: string }>} */
    const indexedForbidConfigs = {};

    forbidConfiguration.forEach((item) => {
      if (typeof item === 'string') {
        indexedForbidConfigs[item] = { element: item };
      } else {
        indexedForbidConfigs[item.element] = item;
      }
    });

    function reportIfForbidden(element, node) {
      if (has(indexedForbidConfigs, element)) {
        const message = indexedForbidConfigs[element].message;

        report(
          context,
          message ? messages.forbiddenElement_message : messages.forbiddenElement,
          message ? 'forbiddenElement_message' : 'forbiddenElement',
          {
            node,
            data: {
              element,
              message,
            },
          }
        );
      }
    }

    return {
      JSXOpeningElement(node) {
        reportIfForbidden(getText(context, node.name), node.name);
      },

      CallExpression(node) {
        if (!isCreateElement(context, node)) {
          return;
        }

        const argument = node.arguments[0];
        if (!argument) {
          return;
        }

        if (argument.type === 'Identifier' && /^[A-Z_]/.test(argument.name)) {
          reportIfForbidden(argument.name, argument);
        } else if (argument.type === 'Literal' && /^[a-z][^.]*$/.test(String(argument.value))) {
          reportIfForbidden(argument.value, argument);
        } else if (argument.type === 'MemberExpression') {
          reportIfForbidden(getText(context, argument), argument);
        }
      },
    };
  },
};
