// eslint-plugin-react@7.37.5/lib/rules/jsx-no-target-blank.js
/**
 * @fileoverview Forbid target='_blank' attribute
 * @author Kevin Miller
 */

'use strict';

const includes = require('array-includes');
const docsUrl = require('../util/docsUrl');
const linkComponentsUtil = require('../util/linkComponents');
const report = require('../util/report');

// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------

function findLastIndex(arr, condition) {
  for (let i = arr.length - 1; i >= 0; i -= 1) {
    if (condition(arr[i])) {
      return i;
    }
  }

  return -1;
}

function attributeValuePossiblyBlank(attribute) {
  if (!attribute || !attribute.value) {
    return false;
  }
  const value = attribute.value;
  if (value.type === &apos;Literal&apos;) {
    return typeof value.value === &apos;string&apos; && value.value.toLowerCase() === &apos;_blank&apos;;
  }
  if (value.type === &apos;JSXExpressionContainer&apos;) {
    const expr = value.expression;
    if (expr.type === &apos;Literal&apos;) {
      return typeof expr.value === &apos;string&apos; && expr.value.toLowerCase() === &apos;_blank&apos;;
    }
    if (expr.type === &apos;ConditionalExpression&apos;) {
      if (expr.alternate.type === &apos;Literal&apos; && expr.alternate.value && expr.alternate.value.toLowerCase() === &apos;_blank&apos;) {
        return true;
      }
      if (expr.consequent.type === &apos;Literal&apos; && expr.consequent.value && expr.consequent.value.toLowerCase() === &apos;_blank&apos;) {
        return true;
      }
    }
  }
  return false;
}

function hasExternalLink(node, linkAttributes, warnOnSpreadAttributes, spreadAttributeIndex) {
  const linkIndex = findLastIndex(node.attributes, (attr) => attr.name && includes(linkAttributes, attr.name.name));
  const foundExternalLink = linkIndex !== -1 && ((attr) => attr.value && attr.value.type === &apos;Literal&apos; && /^(?:\w+:|\/\/)/.test(attr.value.value))(
    node.attributes[linkIndex]);
  return foundExternalLink || (warnOnSpreadAttributes && linkIndex < spreadAttributeIndex);
}

function hasDynamicLink(node, linkAttributes) {
  const dynamicLinkIndex = findLastIndex(node.attributes, (attr) => attr.name
    && includes(linkAttributes, attr.name.name)
    && attr.value
    && attr.value.type === &apos;JSXExpressionContainer&apos;);
  if (dynamicLinkIndex !== -1) {
    return true;
  }
}

/**
 * Get the string(s) from a value
 * @param {ASTNode} value The AST node being checked.
 * @param {ASTNode} targetValue The AST node being checked.
 * @returns {string | string[] | null} The string value, or null if not a string.
 */
function getStringFromValue(value, targetValue) {
  if (value) {
    if (value.type === &apos;Literal&apos;) {
      return value.value;
    }
    if (value.type === &apos;JSXExpressionContainer&apos;) {
      if (value.expression.type === &apos;TemplateLiteral&apos;) {
        return value.expression.quasis[0].value.cooked;
      }
      const expr = value.expression;
      if (expr && expr.type === &apos;ConditionalExpression&apos;) {
        const relValues = [expr.consequent.value, expr.alternate.value];
        if (targetValue.type === &apos;JSXExpressionContainer&apos; && targetValue.expression && targetValue.expression.type === &apos;ConditionalExpression&apos;) {
          const targetTestCond = targetValue.expression.test.name;
          const relTestCond = value.expression.test.name;
          if (targetTestCond === relTestCond) {
            const targetBlankIndex = [targetValue.expression.consequent.value, targetValue.expression.alternate.value].indexOf(&apos;_blank&apos;);
            return relValues[targetBlankIndex];
          }
        }
        return relValues;
      }
      return expr.value;
    }
  }
  return null;
}

function hasSecureRel(node, allowReferrer, warnOnSpreadAttributes, spreadAttributeIndex) {
  const relIndex = findLastIndex(node.attributes, (attr) => (attr.type === &apos;JSXAttribute&apos; && attr.name.name === &apos;rel&apos;));
  const targetIndex = findLastIndex(node.attributes, (attr) => (attr.type === &apos;JSXAttribute&apos; && attr.name.name === &apos;target&apos;));
  if (relIndex === -1 || (warnOnSpreadAttributes && relIndex < spreadAttributeIndex)) {
    return false;
  }

  const relAttribute = node.attributes[relIndex];
  const targetAttributeValue = node.attributes[targetIndex] && node.attributes[targetIndex].value;
  const value = getStringFromValue(relAttribute.value, targetAttributeValue);
  return [].concat(value).every((item) => {
    const tags = typeof item === &apos;string&apos; ? item.toLowerCase().split(&apos; &apos;) : false;
    const noreferrer = tags && tags.indexOf(&apos;noreferrer&apos;) >= 0;
    if (noreferrer) {
      return true;
    }
    const noopener = tags && tags.indexOf(&apos;noopener&apos;) >= 0;
    return allowReferrer && noopener;
  });
}

const messages = {
  noTargetBlankWithoutNoreferrer: &apos;Using target=&quot;_blank&quot; without rel=&quot;noreferrer&quot; (which implies rel=&quot;noopener&quot;) is a security risk in older browsers: see https://mathiasbynens.github.io/rel-noopener/#recommendations&apos;,
  noTargetBlankWithoutNoopener: &apos;Using target=&quot;_blank&quot; without rel=&quot;noreferrer&quot; or rel=&quot;noopener&quot; (the former implies the latter and is preferred due to wider support) is a security risk: see https://mathiasbynens.github.io/rel-noopener/#recommendations&apos;,
};

/** @type {import(&apos;eslint&apos;).Rule.RuleModule} */
module.exports = {
  meta: {
    fixable: &apos;code&apos;,
    docs: {
      description: &apos;Disallow `target=&quot;_blank&quot;` attribute without `rel=&quot;noreferrer&quot;`&apos;,
      category: &apos;Best Practices&apos;,
      recommended: true,
      url: docsUrl(&apos;jsx-no-target-blank&apos;),
    },

    messages,

    schema: [{
      type: &apos;object&apos;,
      properties: {
        allowReferrer: {
          type: &apos;boolean&apos;,
        },
        enforceDynamicLinks: {
          enum: [&apos;always&apos;, &apos;never&apos;],
        },
        warnOnSpreadAttributes: {
          type: &apos;boolean&apos;,
        },
        links: {
          type: &apos;boolean&apos;,
          default: true,
        },
        forms: {
          type: &apos;boolean&apos;,
          default: false,
        },
      },
      additionalProperties: false,
    }],
  },

  create(context) {
    const configuration = Object.assign(
      {
        allowReferrer: false,
        warnOnSpreadAttributes: false,
        links: true,
        forms: false,
      },
      context.options[0]
    );
    const allowReferrer = configuration.allowReferrer;
    const warnOnSpreadAttributes = configuration.warnOnSpreadAttributes;
    const enforceDynamicLinks = configuration.enforceDynamicLinks || &apos;always&apos;;
    const linkComponents = linkComponentsUtil.getLinkComponents(context);
    const formComponents = linkComponentsUtil.getFormComponents(context);

    return {
      JSXOpeningElement(node) {
        const targetIndex = findLastIndex(node.attributes, (attr) => attr.name && attr.name.name === &apos;target&apos;);
        const spreadAttributeIndex = findLastIndex(node.attributes, (attr) => (attr.type === &apos;JSXSpreadAttribute&apos;));

        if (linkComponents.has(node.name.name)) {
          if (!attributeValuePossiblyBlank(node.attributes[targetIndex])) {
            const hasSpread = spreadAttributeIndex >= 0;

            if (warnOnSpreadAttributes && hasSpread) {
              // continue to check below
            } else if ((hasSpread && targetIndex < spreadAttributeIndex) || !hasSpread || !warnOnSpreadAttributes) {
              return;
            }
          }

          const linkAttributes = linkComponents.get(node.name.name);
          const hasDangerousLink = hasExternalLink(node, linkAttributes, warnOnSpreadAttributes, spreadAttributeIndex)
            || (enforceDynamicLinks === 'always' && hasDynamicLink(node, linkAttributes));
          if (hasDangerousLink && !hasSecureRel(node, allowReferrer, warnOnSpreadAttributes, spreadAttributeIndex)) {
            const messageId = allowReferrer ? 'noTargetBlankWithoutNoopener' : 'noTargetBlankWithoutNoreferrer';
            const relValue = allowReferrer ? 'noopener' : 'noreferrer';
            report(context, messages[messageId], messageId, {
              node,
              fix(fixer) {
                // eslint 5 uses `node.attributes`; eslint 6+ uses `node.parent.attributes`
                const nodeWithAttrs = node.parent.attributes ? node.parent : node;
                // eslint 5 does not provide a `name` property on JSXSpreadElements
                const relAttribute = nodeWithAttrs.attributes.find((attr) => attr.name && attr.name.name === &apos;rel&apos;);

                if (targetIndex < spreadAttributeIndex || (spreadAttributeIndex >= 0 && !relAttribute)) {
                  return null;
                }

                if (!relAttribute) {
                  return fixer.insertTextAfter(nodeWithAttrs.attributes.slice(-1)[0], ` rel=&quot;${relValue}&quot;`);
                }

                if (!relAttribute.value) {
                  return fixer.insertTextAfter(relAttribute, `=&quot;${relValue}&quot;`);
                }

                if (relAttribute.value.type === &apos;Literal&apos;) {
                  const parts = relAttribute.value.value
                    .split(&apos;noreferrer&apos;)
                    .filter(Boolean);
                  return fixer.replaceText(relAttribute.value, `&quot;${parts.concat(&apos;noreferrer&apos;).join(&apos; &apos;)}&quot;`);
                }

                if (relAttribute.value.type === &apos;JSXExpressionContainer&apos;) {
                  if (relAttribute.value.expression.type === &apos;Literal&apos;) {
                    if (typeof relAttribute.value.expression.value === &apos;string&apos;) {
                      const parts = relAttribute.value.expression.value
                        .split(&apos;noreferrer&apos;)
                        .filter(Boolean);
                      return fixer.replaceText(relAttribute.value.expression, `&quot;${parts.concat(&apos;noreferrer&apos;).join(&apos; &apos;)}&quot;`);
                    }

                    // for undefined, boolean, number, symbol, bigint, and null
                    return fixer.replaceText(relAttribute.value, &apos;&quot;noreferrer&quot;&apos;);
                  }
                }

                return null;
              },
            });
          }
        }
        if (formComponents.has(node.name.name)) {
          if (!attributeValuePossiblyBlank(node.attributes[targetIndex])) {
            const hasSpread = spreadAttributeIndex >= 0;

            if (warnOnSpreadAttributes && hasSpread) {
              // continue to check below
            } else if (
              (hasSpread && targetIndex < spreadAttributeIndex)
              || !hasSpread
              || !warnOnSpreadAttributes
            ) {
              return;
            }
          }

          if (!configuration.forms || hasSecureRel(node)) {
            return;
          }

          const formAttributes = formComponents.get(node.name.name);

          if (
            hasExternalLink(node, formAttributes)
            || (enforceDynamicLinks === 'always' && hasDynamicLink(node, formAttributes))
          ) {
            const messageId = allowReferrer ? 'noTargetBlankWithoutNoopener' : 'noTargetBlankWithoutNoreferrer';
            report(context, messages[messageId], messageId, {
              node,
            });
          }
        }
      },
    };
  },
};
