// eslint-plugin-react@7.37.5/lib/rules/no-unescaped-entities.js
/**
 * @fileoverview HTML special characters should be escaped.
 * @author Patrick Hayes
 */

'use strict';

const docsUrl = require('../util/docsUrl');
const getSourceCode = require('../util/eslint').getSourceCode;
const jsxUtil = require('../util/jsx');
const report = require('../util/report');
const getMessageData = require('../util/message');

// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------

// NOTE: '<' and '{' are also problematic characters, but they do not need
// to be included here because it is a syntax error when these characters are
// included accidentally.
const DEFAULTS = [{
  char: '>&apos;,
  alternatives: [&apos;&gt;&apos;],
}, {
  char: &apos;&quot;&apos;,
  alternatives: [&apos;&quot;&apos;, &apos;&ldquo;&apos;, &apos;&#34;&apos;, &apos;&rdquo;&apos;],
}, {
  char: &apos;\&apos;&apos;,
  alternatives: [&apos;&apos;&apos;, &apos;&lsquo;&apos;, &apos;&#39;&apos;, &apos;&rsquo;&apos;],
}, {
  char: &apos;}&apos;,
  alternatives: [&apos;&#125;&apos;],
}];

const messages = {
  unescapedEntity: &apos;HTML entity, `{{entity}}` , must be escaped.&apos;,
  unescapedEntityAlts: &apos;`{{entity}}` can be escaped with {{alts}}.&apos;,
  replaceWithAlt: &apos;Replace with `{{alt}}`.&apos;,
};

/** @type {import(&apos;eslint&apos;).Rule.RuleModule} */
module.exports = {
  meta: {
    hasSuggestions: true,
    docs: {
      description: &apos;Disallow unescaped HTML entities from appearing in markup&apos;,
      category: &apos;Possible Errors&apos;,
      recommended: true,
      url: docsUrl(&apos;no-unescaped-entities&apos;),
    },

    messages,

    schema: [{
      type: &apos;object&apos;,
      properties: {
        forbid: {
          type: &apos;array&apos;,
          items: {
            anyOf: [{
              type: &apos;string&apos;,
            }, {
              type: &apos;object&apos;,
              properties: {
                char: {
                  type: &apos;string&apos;,
                },
                alternatives: {
                  type: &apos;array&apos;,
                  uniqueItems: true,
                  items: {
                    type: &apos;string&apos;,
                  },
                },
              },
            }],
          },
        },
      },
      additionalProperties: false,
    }],
  },

  create(context) {
    function reportInvalidEntity(node) {
      const configuration = context.options[0] || {};
      const entities = configuration.forbid || DEFAULTS;

      // HTML entities are already escaped in node.value (as well as node.raw),
      // so pull the raw text from getSourceCode(context)
      for (let i = node.loc.start.line; i <= node.loc.end.line; i++) {
        let rawLine = getSourceCode(context).lines[i - 1];
        let start = 0;
        let end = rawLine.length;
        if (i === node.loc.start.line) {
          start = node.loc.start.column;
        }
        if (i === node.loc.end.line) {
          end = node.loc.end.column;
        }
        rawLine = rawLine.slice(start, end);
        for (let j = 0; j < entities.length; j++) {
          for (let index = 0; index < rawLine.length; index++) {
            const c = rawLine[index];
            if (typeof entities[j] === 'string') {
              if (c === entities[j]) {
                report(context, messages.unescapedEntity, 'unescapedEntity', {
                  node,
                  loc: { line: i, column: start + index },
                  data: {
                    entity: entities[j],
                  },
                });
              }
            } else if (c === entities[j].char) {
              report(context, messages.unescapedEntityAlts, 'unescapedEntityAlts', {
                node,
                loc: { line: i, column: start + index },
                data: {
                  entity: entities[j].char,
                  alts: entities[j].alternatives.map((alt) => `\`${alt}\``).join(', '),
                },
                suggest: entities[j].alternatives.map((alt) => Object.assign(
                  getMessageData('replaceWithAlt', messages.replaceWithAlt),
                  {
                    data: { alt },
                    fix(fixer) {
                      const lineToChange = i - node.loc.start.line;

                      const newText = node.raw.split('\n').map((line, idx) => {
                        if (idx === lineToChange) {
                          return line.slice(0, index) + alt + line.slice(index + 1);
                        }

                        return line;
                      }).join('\n');

                      return fixer.replaceText(node, newText);
                    },
                  }
                )),
              });
            }
          }
        }
      }
    }

    return {
      'Literal, JSXText'(node) {
        if (jsxUtil.isJSX(node.parent)) {
          reportInvalidEntity(node);
        }
      },
    };
  },
};
