// @hookform/resolvers@3.10.0/computed-types/src/__tests__/__fixtures__/data.ts
import Schema, { Type, string, number, array, boolean } from 'computed-types';
import { Field, InternalFieldName } from 'react-hook-form';

export const schema = Schema({
  username: string.regexp(/^\w+$/).min(3).max(30),
  password: string
    .regexp(new RegExp('.*[A-Z].*'), 'One uppercase character')
    .regexp(new RegExp('.*[a-z].*'), 'One lowercase character')
    .regexp(new RegExp('.*\\d.*'), 'One number')
    .regexp(
      new RegExp('.*[`~<>?,./!@#$%^&*()\\-_+=&quot;\&apos;|{}\\[\\];:\\\\].*&apos;),
      &apos;One special character&apos;,
    )
    .min(8, &apos;Must be at least 8 characters in length&apos;),
  repeatPassword: string,
  accessToken: Schema.either(string, number).optional(),
  birthYear: number.min(1900).max(2013).optional(),
  email: string
    .regexp(/^[\w-\.]+@([\w-]+\.)+[\w-]{2,4}$/)
    .error(&apos;Incorrect email&apos;),
  tags: array.of(string),
  enabled: boolean,
  like: array
    .of({
      id: number,
      name: string.min(4).max(4),
    })
    .optional(),
  address: Schema({
    city: string.min(3, &apos;Is required&apos;),
    zipCode: string
      .min(5, &apos;Must be 5 characters long&apos;)
      .max(5, &apos;Must be 5 characters long&apos;),
  }),
});

export const validData: Type<typeof schema> = {
  username: &apos;Doe&apos;,
  password: &apos;Password123_&apos;,
  repeatPassword: &apos;Password123_&apos;,
  accessToken: &apos;accessToken&apos;,
  birthYear: 2000,
  email: &apos;john@doe.com&apos;,
  tags: [&apos;tag1&apos;, &apos;tag2&apos;],
  enabled: true,
  like: [
    {
      id: 1,
      name: &apos;name&apos;,
    },
  ],
  address: {
    city: &apos;Awesome city&apos;,
    zipCode: &apos;12345&apos;,
  },
};

export const invalidData = {
  password: &apos;___&apos;,
  email: &apos;&apos;,
  birthYear: &apos;birthYear&apos;,
  like: [{ id: &apos;z&apos; }],
  address: {
    city: &apos;&apos;,
    zipCode: &apos;123&apos;,
  },
};

export const fields: Record<InternalFieldName, Field['_f']> = {
  username: {
    ref: { name: 'username' },
    name: 'username',
  },
  password: {
    ref: { name: 'password' },
    name: 'password',
  },
  email: {
    ref: { name: 'email' },
    name: 'email',
  },
  birthday: {
    ref: { name: 'birthday' },
    name: 'birthday',
  },
};
