// @hookform/resolvers@3.10.0/io-ts/src/__tests__/__fixtures__/data.ts
import * as t from 'io-ts';
import * as tt from 'io-ts-types';

import { Field, InternalFieldName } from 'react-hook-form';

export const schema = t.intersection([
  t.type({
    username: tt.NonEmptyString,
    password: tt.NonEmptyString,
    accessToken: tt.UUID,
    birthYear: t.number,
    email: t.string,
    tags: t.array(
      t.type({
        name: t.string,
      }),
    ),
    luckyNumbers: t.array(t.number),
    enabled: t.boolean,
    animal: t.union([
      t.string,
      t.number,
      t.literal('bird'),
      t.literal('snake'),
    ]),
    vehicles: t.array(
      t.union([
        t.type({
          type: t.literal('car'),
          brand: t.string,
          horsepower: t.number,
        }),
        t.type({
          type: t.literal('bike'),
          speed: t.number,
        }),
      ]),
    ),
  }),
  t.partial({
    like: t.array(
      t.type({
        id: tt.withMessage(
          t.number,
          (i) => `this id is very important but you passed: ${typeof i}(${i})`,
        ),
        name: t.string,
      }),
    ),
  }),
]);

interface Data {
  username: string;
  password: string;
  accessToken: string;
  birthYear?: number;
  luckyNumbers: number[];
  email?: string;
  animal: string | number;
  tags: { name: string }[];
  enabled: boolean;
  like: { id: number; name: string }[];
  vehicles: Array<
    | { type: 'car'; brand: string; horsepower: number }
    | { type: 'bike'; speed: number }
  >;
}

export const validData: Data = {
  username: &apos;Doe&apos;,
  password: &apos;Password123&apos;,
  accessToken: &apos;c2883927-5178-4ad1-bbee-07ba33a5de19&apos;,
  birthYear: 2000,
  email: &apos;john@doe.com&apos;,
  tags: [{ name: &apos;test&apos; }],
  enabled: true,
  luckyNumbers: [17, 5],
  animal: &apos;cat&apos;,
  like: [
    {
      id: 1,
      name: &apos;name&apos;,
    },
  ],
  vehicles: [{ type: &apos;car&apos;, brand: &apos;BMW&apos;, horsepower: 150 }],
};

export const invalidData = {
  username: &apos;test&apos;,
  password: &apos;Password123&apos;,
  repeatPassword: &apos;Password123&apos;,
  birthYear: 2000,
  accessToken: &apos;1015d809-e99d-41ec-b161-981a3c243df8&apos;,
  email: &apos;john@doe.com&apos;,
  tags: [{ name: &apos;test&apos; }],
  enabled: true,
  animal: [&apos;dog&apos;],
  luckyNumbers: [1, 2, &apos;3&apos;],
  like: [
    {
      id: &apos;1&apos;,
      name: &apos;name&apos;,
    },
  ],
  vehicles: [
    { type: &apos;car&apos;, brand: &apos;BMW&apos;, horsepower: 150 },
    { type: &apos;car&apos;, brand: &apos;Mercedes&apos; },
  ],
};

export const fields: Record<InternalFieldName, Field['_f']> = {
  username: {
    ref: { name: 'username' },
    name: 'username',
  },
  password: {
    ref: { name: 'password' },
    name: 'password',
  },
  email: {
    ref: { name: 'email' },
    name: 'email',
  },
  birthday: {
    ref: { name: 'birthday' },
    name: 'birthday',
  },
};
